---
title: "alcohol_LOAD_two_sample_mr"
author: "Kyle Abraham Campbell"
date: "9/26/2021"
output: html_document
---

```{r setup, include=FALSE}
library(gtsummary)
library(here)
library(knitr)
library(parameters)
library(readxl)
library(tidymodels)
library(tidyverse)

#library(remotes)
#remotes::install_github("MRCIEU/TwoSampleMR")
library(TwoSampleMR)

knitr::opts_chunk$set(echo = TRUE)
```

Helper function to output simple tables of TwoSampleMR package function results tables
```{r}
quick_2smr_kable <- function(x) {
  return <-
    x %>%
    select(!c(id.exposure, id.outcome)) %>%
    kable(digits = 2)
  return(return)
}
```

From Supplemental Table 5 from Liu et al., 2019. Same Alcohol consumption IVs used in Rosoff et al., 2020. A chr17 SNP rsid is labelled missing. Look into recovering with SNP database
```{r full_GSCAN_Liu_2019_res}
gscan.dpw <- read_excel(here("data", "gscan_dpw_full.xlsx"))
head(gscan.dpw)
```

Phenotype variation explained (PVE) is a substitute for R^2 in summary GWAS data (h^2, heritability). Calculate F statistic from that, total F-stat based on mzfu's code
https://stats.stackexchange.com/questions/56881/whats-the-relationship-between-r2-and-f-test
Teslovich TM, Musunuru K, Smith AV, Edmondson AC, Stylianou IM, Koseki M, et al. Biological,
clinical and population relevance of 95 loci for blood lipids. Nature. 2010;466:707â€“713.

```{r}
gscan.dpw.formatted <-
  gscan.dpw %>%
  rename_with(tolower) %>%
  rename(
    SNP = rsid,
    effect_allele = "alternate allele",
    other_allele = "reference allele",
    eaf = "alternate allele frequency",
    pval = pvalue,
    samplesize = "effective n"
  ) %>%
  mutate(pve = (2*beta^2*eaf*(1-eaf))/(2*beta^2*eaf*(1-eaf)+se^2*2*n*eaf*(1-eaf))) %>%
  mutate(fstat_snp = ((n-1-1)*pve)/1/(1-pve)^2) %>%
  mutate(fstat = ((n-nrow(gscan.dpw)-1)*(sum(pve))/nrow(gscan.dpw)/(1-sum(pve))^2))

names <- colnames(gscan.dpw.formatted)[1:12]
gscan.dpw.formatted %>% select(names, pve, fstat_snp, fstat)
```

```{r}
dat.exp <- format_data(gscan.dpw.formatted)
# By default, clumps all index and secondary SNPs at a 10000kb resolution with r^2 > .001
dat.exp.clump <- clump_data(dat.exp)
```

## Extract summary GWASs asnd harmonize

Extract Alcohol IV from 2019 GSCAN Consortium paper in the IEU GWAS Catalog (ID: ieu-b-73), extract outcome data - Late Onset Alzheimer's Disease (Kunkle et al., 2019) from the IEU GWAS Catalog (ID: ieu-b-2), and harmonize

```{r}
alc_exp_dat <- dat.exp.clump
#alc_exp_dat <- extract_instruments(outcomes = 'ieu-b-73')
```

```{r}
load_out_dat <- extract_outcome_data(snps = alc_exp_dat$SNP, outcomes = 'ebi-a-GCST005069')
dat <- harmonise_data(alc_exp_dat, load_out_dat)
```
```{r}
dat <- dat %>%
  filter(!(SNP %in% c("rs28929474", "rs1260326")))
```


Select which 2SMR models to run
```{r}
methods.list <-
  mr_method_list() %>%
  filter(use_by_default == T | 
           #heterogeneity_test == T | 
           obj == "mr_ivw_mre") %>%
  pull(obj)

methods.hetero <-
  mr_method_list() %>%
  filter(heterogeneity_test == T) %>% # & 
           #(use_by_default == T | 
              #obj == "mr_ivw_mre")) %>% 
  pull(obj)
```

Run 2SMR models
```{r}
res <- mr(dat, method_list = methods.list)
```

```{r}
#res.or <- generate_odds_ratios(res)
#quick_2smr_kable(res.or)
quick_2smr_kable(res)
```

No significant results with any of the default MR tests in TwoSampleMR. Simple mode prioritizes the collection of moderate inverse effect SNP clusters. Weighting in weighted mode attenuates this effect.

```{r}
p1 <- mr_scatter_plot(res, dat)
p1[[1]]
```

Heterogeneity in effects, suggesting violating of IV or modelling assumptions, likely due to     horizontal pleiotropy. Further discussion here: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6659376/
```{r}
mr_heterogeneity(dat, method_list = methods.hetero) %>% quick_2smr_kable
```

Egger intercept is not different than 0, suggests no bias due to directional horizontal pleiotropy
```{r}
mr_pleiotropy_test(dat) %>% quick_2smr_kable
```

rs56030824 appears to be an outlier here and is confirmed so with MR-PRESSO below.
```{r}
res_single <- mr_singlesnp(dat)
p2 <- mr_forest_plot(res_single)
p2[[1]]
```

LOO suggests there may be a few extreme outliers.
```{r}
res_loo <- mr_leaveoneout(dat)
p3 <- mr_leaveoneout_plot(res_loo)
p3[[1]]
```

rs7452999 is an outlier here, too.
```{r}
res_single <- mr_singlesnp(dat)
p4 <- mr_funnel_plot(res_single)
p4[[1]]
```

Steiger test of causal direction suggests proposed direction is accurate.
```{r}
out <- directionality_test(dat)
quick_2smr_kable(out)
```
