---
title: "alcohol_LOAD_two_sample_mr"
author: "Kyle Abraham Campbell"
date: "9/26/2021"
output: html_document
---

```{r setup, include=FALSE}
library(devtools)
library(DiagrammeR)
library(DiagrammeRsvg)
library(glue)
library(gtsummary)
library(here)
library(knitr)
library(parameters)
library(readxl)
library(tidymodels)
library(tidyverse)

#library(remotes)
#remotes::install_github("qingyuanzhao/mr.raps")
library(mr.raps)
#devtools::install_github("WSpiller/RadialMR")
library(RadialMR)
devtools::install_github("explodecomputer/tryx")
library(tryx)
#remotes::install_github("MRCIEU/TwoSampleMR")
library(TwoSampleMR)

knitr::opts_chunk$set(echo = TRUE)
```

Helper function to output simple tables of TwoSampleMR package function results tables
```{r}
quick_2smr_kable <- function(x) {
  return <-
    x %>%
    select(!c(id.exposure, id.outcome, exposure, outcome)) %>%
    kable(digits = 2)
  return(return)
}
```

# Two Sample MR of Alcohol and Dementia focusing on PGSCAN and IGAP consortia GWASs {.tabset}

## Top hit GSCAN GWAS results for weekly alcohol consumption (includes 23andMe)

From Supplemental Table 5 from Liu et al., 2019. Same Alcohol consumption IVs used in Rosoff et al., 2020. A chr17 SNP rsid is labelled missing. Look into recovering with SNP database
```{r full_GSCAN_Liu_2019_res}
#TODO check if chr17 SNP can be identified with RSID
gscan.dpw <- read_excel(here("data", "gscan_dpw_full.xlsx"))
head(gscan.dpw)
paste0(dim(gscan.dpw)[1], " significant SNPs for genetic instrument consideration")
#write_tsv(gscan.dpw, file = here("data", "gscan_dpw_hits.tsv"))
```

Export a list of variants split by chromosome for input into bcftools to subset HRS imputed_geno .vcfs
```{r}
gscan.chr.pos <- gscan.dpw %>%
  select(Chr, Position) %>%
  #mutate(region = paste0(Chr, ":", Position)) %>%
  group_by(Chr) %>%
  group_split()
chr <- map(gscan.chr.pos, ~ .x$Chr[1])
names(gscan.chr.pos) <- chr

gscan.chr.pos.df <- map2(.x = gscan.chr.pos,
                         .y = chr,
                         ~ .x %>%
                           as.data.frame %>%
                           write_tsv(file = here("data", "gscan_chr_hits", paste0("chr.", .y, ".gscan.dpw.chr.pos.tsv")),
                                     col_names = FALSE))

#write_csv(x = gscan.chr.pos, file = here("data", "gscan_dpw_chr_pos.csv"), col_names = FALSE)
```

Phenotype variation explained (PVE) is a substitute for R^2 in summary GWAS data (h^2, heritability). Calculate F statistic from that, total F-stat based on mzfu's code.
https://stats.stackexchange.com/questions/56881/whats-the-relationship-between-r2-and-f-test
Teslovich TM, Musunuru K, Smith AV, Edmondson AC, Stylianou IM, Koseki M, et al. Biological,
clinical and population relevance of 95 loci for blood lipids. Nature. 2010;466:707â€“713.

```{r}
gscan.dpw.formatted <-
  gscan.dpw %>%
  rename_with(tolower) %>%
  rename(
    SNP = rsid,
    effect_allele = "alternate allele",
    other_allele = "reference allele",
    eaf = "alternate allele frequency",
    pval = pvalue,
    samplesize = "effective n"
  ) %>%
  mutate(pve = (2*beta^2*eaf*(1-eaf))/(2*beta*eaf*(1-eaf)+se^2*2*n*eaf*(1-eaf))) %>%
  mutate(fstat_snp = ((n-1-1)*pve)/1/(1-pve)^2) %>%
  mutate(fstat = ((n-nrow(gscan.dpw)-1)*(sum(pve))/nrow(gscan.dpw)/(1-sum(pve))^2)) %>%
  mutate(pval = ifelse(pval == "<2.2e-308", 2.2e-307, pval))

names <- colnames(gscan.dpw.formatted)[1:12]
#gscan.dpw.formatted %>% select(names, pve, fstat_snp, fstat) %>% View
```

```{r}
ggplot(data = gscan.dpw.formatted, mapping = (aes( x = fstat_snp))) +
  geom_histogram() +
  theme_bw()

ggplot(data = gscan.dpw.formatted, mapping = (aes( x = fstat_snp))) +
  geom_histogram() +
  theme_bw() +
  xlim(0,100)
```

Format data for 2SMR input and clump
```{r}
dat.exp <- format_data(gscan.dpw.formatted)
snps.in.eur <- ieugwasr::ld_reflookup(dat.exp$SNP, pop = "EUR")
```

There are two introns not included in the EUR reference (rs11739827; 5:166303321-167303321, .; 17:43159975-45273783) (bi-allelic SNPs with MAF > 0.01); the other 26 SNPs were excluded as correlated; There were 18 non-sentinel SNPs in GSCAN DPW results, but GSCAN GWAS used 1MB window while we use 10 here
```{r}
not.in.eur.ref <- gscan.dpw.formatted %>% filter(!(SNP %in% snps.in.eur)) #%>% View()
not.in.eur.ref
```

```{r}
# By default, clumps all index and secondary SNPs at a 10000kb resolution with r^2 > .001
dat.exp.clump <- clump_data(dat.exp)
```

Recalculate fstats after clumping
```{r}
instrument.ids <- dat.exp.clump %>% pull(SNP)
instrument.info <- gscan.dpw.formatted %>%
  filter(SNP %in% instrument.ids) %>%
  mutate(pve = (2*beta^2*eaf*(1-eaf))/(2*beta^2*eaf*(1-eaf)+se^2*2*n*eaf*(1-eaf))) %>%
  mutate(fstat_snp = ((n-1-1)*pve)/1/(1-pve)^2) %>%
  mutate(fstat = ((n-nrow(gscan.dpw)-1)*(sum(pve))/nrow(gscan.dpw)/(1-sum(pve))^2))


ggplot(data = instrument.info, mapping = (aes( x = fstat_snp))) +
  geom_histogram() +
  theme_bw()

ggplot(data = instrument.info, mapping = (aes( x = fstat_snp))) +
  geom_histogram() +
  theme_bw() +
  xlim(0,100)
```

```{r}
alc_exp_dat <- dat.exp.clump
independent.snps <- alc_exp_dat %>% pull(SNP) 
#saveRDS(instrument.info %>% filter(SNP %in% independent.snps), here("data", "analytic", "liu_independent_instruments.rda"))
instrument.info %>% filter(SNP %in% independent.snps) %>% pull(locus)
instrument.info %>% filter(SNP %in% independent.snps) %>% pull(locus) %>% str_extract(pattern = "\\d*:\\d*")
```

```{r}
# rs17177078 not included due to not being present in the outcome GWAS and no proxy SNP found?
load_out_dat <- extract_outcome_data(snps = alc_exp_dat$SNP, outcomes = 'ieu-b-2')
dat <- harmonise_data(alc_exp_dat, load_out_dat)
```


Recalculate fstats after harmonization
```{r}
instrument.ids <- dat %>% filter(mr_keep == T) %>% pull(SNP)
instrument.info <- gscan.dpw.formatted %>%
  filter(SNP %in% instrument.ids) %>%
  mutate(pve = (2*beta^2*eaf*(1-eaf))/(2*beta^2*eaf*(1-eaf)+se^2*2*n*eaf*(1-eaf))) %>%
  mutate(fstat_snp = ((n-1-1)*pve)/1/(1-pve)^2) %>%
  mutate(fstat = ((n-nrow(gscan.dpw)-1)*(sum(pve))/nrow(gscan.dpw)/(1-sum(pve))^2))


ggplot(data = instrument.info, mapping = (aes( x = fstat_snp))) +
  geom_histogram() +
  theme_bw()

ggplot(data = instrument.info, mapping = (aes( x = fstat_snp))) +
  geom_histogram() +
  theme_bw() +
  xlim(0,100)
```

```{r}
#View(mr_method_list())
```

## Instrument Inclusion/exclusion

Code to create the inclusion/exclusion flowchart 
```{r}
gscan.hits <- glue("Candidate Instruments - Drinks per week GWAS hits
          n = {dim(gscan.dpw.formatted)[1]}")

gscan.hits.clumped <- glue("Independent index candidate instruments, 10MB window, r^2 < 0.001
          n = {dim(dat.exp.clump)[1]}")
gscan.hits.clumped.excluded <- glue("Non-independent candidate instruments
          n = {dim(gscan.dpw.formatted)[1] - dim(dat.exp.clump)[1] - dim(not.in.eur.ref)[1]}
          Instruments not contained in the 1000 Genomes EUR linkage disequilibrium reference panel
          n = {dim(not.in.eur.ref)[1]}")

gscan.hits.clumped.harmonized <- glue("Harmonized exposure and outcome instruments
          n = {dim(dat %>% filter(mr_keep) == T)[1]}")
gscan.hits.clumped.harmonized.excluded <- glue("Exposure instrument or proxy not found in outcome data
          n = {dim(alc_exp_dat)[1] - dim(load_out_dat)[1]}
          Non-inferable palindromic instruments
          n = {dim(load_out_dat)[1] - dim(dat %>% filter(mr_keep) == T)[1]}")
```

```{r}
graph <- grViz("
digraph cohort_flow_chart {
node [fontname = Helvetica, fontsize = 12, shape = box, width = 4]

gscan_hits[label = '@@1-1']

clumped_blank[label = '', width = 0.01, height = 0.01]
clumped_excluded[label = '@@2-2']
clumped[label = '@@2-1']

harmonized_blank[label = '', width = 0.01, height = 0.01]
harmonized_excluded[label = '@@3-1']
harmonized[label = '@@3-2']

{ rank = same; clumped_blank clumped_excluded}
{ rank = same; harmonized_blank harmonized_excluded}

gscan_hits -> clumped_blank [ dir = none ];
clumped_blank -> clumped_excluded [ minlen = 1 ];
clumped_blank -> clumped;

clumped -> harmonized_blank [ dir = none ];
harmonized_blank -> harmonized_excluded [ minlen = 1 ];
harmonized_blank -> harmonized;

}

[1]: gscan.hits
[2]: c(gscan.hits.clumped, gscan.hits.clumped.excluded)
[3]: c(gscan.hits.clumped.harmonized.excluded, gscan.hits.clumped.harmonized)
")
```

```{r, eval = F}
graph %>% export_svg() %>% charToRaw %>% rsvg_pdf(here("results", paste0("instrument_incl_excl_", Sys.Date(), ".pdf")))
```


### Inclusion exlcusion for complementary, empirical 1SMR HRS analysis


```{r}
gscan.hits <- glue("Candidate Instruments - Drinks per week GWAS hits
          n = {dim(gscan.dpw.formatted)[1]}")

gscan.hits.clumped <- glue("Independent index candidate instruments, 10MB window, r^2 < 0.001
          n = {dim(dat.exp.clump)[1]}")
gscan.hits.clumped.excluded <- glue("Non-independent candidate instruments
          n = {dim(gscan.dpw.formatted)[1] - dim(dat.exp.clump)[1] - dim(not.in.eur.ref)[1]}
          Instruments not contained in the 1000 Genomes EUR linkage disequilibrium reference panel
          n = {dim(not.in.eur.ref)[1]}")

gscan.hits.clumped.harmonized <- glue("Analytic exposure instruments
          n = {dim(dat.exp.clump)[1] - 1}")
gscan.hits.clumped.harmonized.excluded <- glue("Minor allele frequency < 2%
          n = 1")
```

```{r}
graph <- grViz("
digraph cohort_flow_chart {
node [fontname = Helvetica, fontsize = 12, shape = box, width = 4]

gscan_hits[label = '@@1-1']

clumped_blank[label = '', width = 0.01, height = 0.01]
clumped_excluded[label = '@@2-2']
clumped[label = '@@2-1']

harmonized_blank[label = '', width = 0.01, height = 0.01]
harmonized_excluded[label = '@@3-1']
harmonized[label = '@@3-2']

{ rank = same; clumped_blank clumped_excluded}
{ rank = same; harmonized_blank harmonized_excluded}

gscan_hits -> clumped_blank [ dir = none ];
clumped_blank -> clumped_excluded [ minlen = 1 ];
clumped_blank -> clumped;

clumped -> harmonized_blank [ dir = none ];
harmonized_blank -> harmonized_excluded [ minlen = 1 ];
harmonized_blank -> harmonized;

}

[1]: gscan.hits
[2]: c(gscan.hits.clumped, gscan.hits.clumped.excluded)
[3]: c(gscan.hits.clumped.harmonized.excluded, gscan.hits.clumped.harmonized)
")
```

```{r, eval = F}
graph %>% export_svg() %>% charToRaw %>% rsvg_pdf(here("results", paste0("instrument_incl_excl_HRS_1smr_", Sys.Date(), ".pdf")))
```


## Selecting and running 2SMR models

Select which 2SMR models to run
```{r}
methods.list <-
  mr_method_list() %>%
    filter(use_by_default == T | 
          heterogeneity_test == T
           ) %>%
  # Dropping mr_raps for now due to error
  #filter(use_by_default == T | 
  #         heterogeneity_test == T | 
  #         obj == "mr_raps"
  #         ) %>%
  filter(!(name %in% c("Unweighted regression", "Maximum Likelihood"))) %>%
  pull(obj)
  

methods.hetero <-
  mr_method_list() %>%
  filter(heterogeneity_test == T) %>% # & 
           #(use_by_default == T | 
              #obj == "mr_ivw_mre")) %>% 
pull(obj)
```

Run 2SMR models
```{r}
res <- mr(dat, method_list = methods.list)

# res <- mr(dat = dat,
#           parameters = list(
#             test_dist = "z",
#             nboot = 1000,
#             Cov = 0,
#             penk = 20,
#             phi = 1,
#             alpha = 0.05,
#             Qthresh = 0.05,
#             over.dispersion = FALSE,
#             loss.function = "huber"),
#           method_list = c("mr_raps"))

#res <- mr(dat, method_list = methods.list)
```

```{r}
res.or <- generate_odds_ratios(res)
res.or
```

```{r format_and_export_primary_2smr_results}
res.or %>%
  dplyr::select(!id.exposure:exposure) %>%
  dplyr::select(!c(nsnp, b, se, lo_ci, up_ci)) %>%
  rename(
    `Estimator` = method,
    `Odds ratio` = or,
    `Lower 95% CI` = or_lci95,
    `Upper 95% CI` = or_uci95,
    `P-value` = pval
  ) %>%
  mutate(across(where(is.numeric), ~ round(.x, digits = 2))) %>%
  filter(Estimator != "IVW radial") %>%
  dplyr::select(Estimator, `Odds ratio`, `Lower 95% CI`, `Upper 95% CI`, `P-value`) %>%
  gt()
```


No significant results with any of the default MR tests in TwoSampleMR. Simple mode prioritizes the collection of moderate inverse effect SNP clusters. Weighting in weighted mode attenuates this effect.

```{r}
p1 <- mr_scatter_plot(res, dat)
p1[[1]]
```

Heterogeneity in effects, suggesting violating of IV or modeling assumptions, likely due to horizontal pleiotropy. Further discussion here: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6659376/. Similar Q statistics for the IVW and MR Egger suggest the Egger is a no better fit than IVW.
```{r}
mr_heterogeneity(dat, method_list = methods.hetero)
res.heterogeneity <- mr_heterogeneity(dat, method_list = methods.hetero)
```

Egger intercept is not different than 0, suggests no bias due to directional horizontal pleiotropy
```{r}
mr_pleiotropy_test(dat)
```

rs56030824 appears to be an outlier here and is confirmed so with MR-PRESSO below. rs1229984 is the ALDH SNP
```{r mr_single_snp_forest_plot}
res_single <- mr_singlesnp(dat)
p2 <- mr_forest_plot(res_single)
p2[[1]]
```

LOO suggests there may be a few extreme outliers.
```{r}
res_loo <- mr_leaveoneout(dat)
p3 <- mr_leaveoneout_plot(res_loo)
p3[[1]]
```

Export genetic instruments for 1-sample MR and PGS analysis
```{r, eval = F}
# Pull SNPs
snp.list <- res_loo %>% pull(SNP)
# Drop the "All" entry
snp.list <- snp.list[-length(snp.list)]
# Export
#saveRDS(snp.list, here("data", "analytic", "2smr_snp_hits_05_23_2022.rda"))
```

rs56030824 is an outlier here, too.
```{r}
res_single <- mr_singlesnp(dat)
p4 <- mr_funnel_plot(res_single)
p4[[1]]
```

Steiger test of causal direction suggests proposed direction is accurate.
```{r Steiger_test}
out <- directionality_test(dat)
out
```

Increase bootstrap to 2000
```{r mr_presso}
res.presso <- run_mr_presso(dat, NbDistribution = 2000, SignifThreshold = 0.05)
```

rs56030824 identified as an outlier. With it removed, estimate moves even closer to the null.
```{r mr_presso_results}
res.presso[[1]]$`Main MR results`
res.presso[[1]]$`MR-PRESSO results`$`Outlier Test`
dat$SNP[52]
```

```{r}
# extract MR-PRESSO outlier-corrected results
presso.main <- res.presso[[1]]$`Main MR results`[2,]

# extract and format MR-PRESSO model results to be compatible with generate_odds_ratios() output from primary analysis
mrpresso.res <- tibble(
  outcome = res.or$outcome[1],
  exposure = res.or$exposure[1],
  method = "MR-PRESSO outlier-corrected",
  nsnp = res.or$nsnp[1] - res.presso[[1]]$`MR-PRESSO results`$`Outlier Test` %>% filter(Pvalue < 0.05) %>% tally %>% as.numeric,
  pval = presso.main$`P-value`,
  or = exp(presso.main$`Causal Estimate`),
  # TODO is this the correct 95% CI calculation; is that Sd SE or Sd?
  or_lci95 = (presso.main$`Causal Estimate` - 1.96*presso.main$Sd) %>% exp,
  or_uci95 = (presso.main$`Causal Estimate` + 1.96*presso.main$Sd) %>% exp
)
```

```{r extract_aldh_results}
res.aldh <- res_single %>% filter(SNP == "rs1229984")
res.aldh.or <- generate_odds_ratios(res.aldh)
res.outlier <- res_single %>% filter(SNP == "rs56030824")
res.outlier.or <- generate_odds_ratios(res.outlier)

aldh.res <- tibble(
  outcome = res.or$outcome[1],
  exposure = paste0(res.or$exposure[1], ", ", res.aldh.or$SNP[1], " - ALDH only"),
  method = "ALDH SNP only",
  nsnp = 1,
  pval = res.aldh.or$p[1],
  or = res.aldh.or$or[1],
  or_lci95 = res.aldh.or$or_lci95[1],
  or_uci95 = res.aldh.or$or_uci95[1]
)

outlier.res <- tibble(
  outcome = res.or$outcome[1],
  exposure = paste0(res.or$exposure[1], ", ", res.outlier.or$SNP[1], " - outlier only"),
  method = "outlier SNP only",
  nsnp = 1,
  pval = res.outlier.or$p[1],
  or = res.outlier.or$or[1],
  or_lci95 = res.outlier.or$or_lci95[1],
  or_uci95 = res.outlier.or$or_uci95[1]
)
```

```{r}
res.or.liu2019.full <- 
  res.or %>%
  select(outcome, exposure, method, nsnp, pval, or, or_lci95, or_uci95) %>%
  rbind(mrpresso.res, aldh.res, outlier.res) %>%
  rename_with(str_to_title) %>%
  rename("Number of SNPs" = Nsnp , "p" = Pval , "OR" = Or, "Lower 95% CI" = Or_lci95, "Upper 95% CI" = Or_uci95) %>%
  mutate(Method = as.factor(Method)) %>%
  mutate(Outcome = "Alzheimer's Disease, Kunkle et al., 2019", Exposure = "Drinks Per Week, Liu et al., 2019") %>%
  filter(!Method %in% c("Unweighted regression", "outlier SNP only"))
```

Forest plot
```{r}
forest <- ggplot(data=res.or.liu2019.full,
                 aes(
                   x=stringr::str_wrap(Method, 20),
                   y=OR,
                   ymin=`Lower 95% CI`,
                   ymax=`Upper 95% CI`,
                   )) +
  xlab("") +
  ylab("Odds Ratio Per log(drinks/week)") +
  geom_hline(yintercept = 1, linetype = 2) +
  geom_pointrange(position=position_dodge(width = 1)) +
  coord_cartesian(ylim = c(0, 2.5)) +
  coord_flip() +
  theme_bw() + 
  theme(text = element_text(size = 18)) 
forest
#ggsave(here("results", "two_smr_alcohol_load.png"))
```

## GSCAN GWAS results for weekly alcohol consumption (excludes 23andMe)

Extract Alcohol IV from 2019 GSCAN Consortium paper in the IEU GWAS Catalog (ID: ieu-b-73), extract outcome data - Late Onset Alzheimer's Disease (Kunkle et al., 2019) from the IEU GWAS Catalog (ID: ieu-b-2), and harmonize

```{r}
# TRYX requires using an MRBase-curated dataset ieu-b-73 instead of the summary statistics that included 23andMe
alc_exp_dat <- extract_instruments(outcomes = 'ieu-b-73')
```

```{r}
load_out_dat <- extract_outcome_data(snps = alc_exp_dat$SNP, outcomes = 'ieu-b-2')
dat <- harmonise_data(alc_exp_dat, load_out_dat)
```

Run 2SMR models
```{r}
res <- mr(dat, method_list = methods.list)
```

```{r}
res.or <- generate_odds_ratios(res)
quick_2smr_kable(res.or)
```

No significant results with any of the default MR tests in TwoSampleMR. Simple mode prioritizes the collection of moderate inverse effect SNP clusters. Weighting in weighted mode attenuates this effect.

```{r}
p1 <- mr_scatter_plot(res, dat)
p1[[1]]
```

Heterogeneity in effects, suggesting violating of IV or modelling assumptions, likely due to horizontal pleiotropy. Further discussion here: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6659376/. Similar Q statistics for the IVW and MR Egger suggest the Egger is a no better fit than IVW.
```{r}
mr_heterogeneity(dat, method_list = methods.hetero) %>% quick_2smr_kable
```

Egger intercept is not different than 0, suggests no bias due to directional horizontal pleiotropy
```{r}
mr_pleiotropy_test(dat) %>% quick_2smr_kable
```

rs4752999 appears to be an outlier here and is confirmed so with MR-PRESSO below. Appears to located near rs56030824, the outlier in the full analysis.
```{r}
res_single <- mr_singlesnp(dat)
p2 <- mr_forest_plot(res_single)
p2[[1]]
```

LOO suggests there may be an outlier.
```{r}
res_loo <- mr_leaveoneout(dat)
p3 <- mr_leaveoneout_plot(res_loo)
p3[[1]]
```

rs4752999 is the outlier here, too.
```{r}
res_single <- mr_singlesnp(dat)
p4 <- mr_funnel_plot(res_single)
p4[[1]]
```

Steiger test of causal direction suggests proposed direction is accurate.
```{r}
out <- directionality_test(dat)
quick_2smr_kable(out)
```

Increase bootstrap to 2000
```{r}
res.presso <- run_mr_presso(dat, NbDistribution = 2000, SignifThreshold = 0.05)
```

rs4752999 identified as an outlier. With it removed, estimate moves even closer to the null.
```{r}
res.presso[[1]]$`Main MR results`
res.presso[[1]]$`MR-PRESSO results`$`Outlier Test`
dat$SNP[21]
```

```{r}
# extract MR-PRESSO outlier-corrected results
presso.main <- res.presso[[1]]$`Main MR results`[2,]

# Extract and format MR-PRESSO model results to be compatible with generate_odds_ratios() output from primary analysis
mrpresso.res <- tibble(
  outcome = res.or$outcome[1],
  exposure = res.or$exposure[1],
  method = "MR-PRESSO outlier-corrected",
  nsnp = res.or$nsnp[1] - res.presso[[1]]$`MR-PRESSO results`$`Outlier Test` %>% filter(Pvalue < 0.05) %>% tally %>% as.numeric,
  pval = presso.main$`P-value`,
  or = exp(presso.main$`Causal Estimate`),
  or_lci95 = (presso.main$`Causal Estimate` - 1.96*presso.main$Sd) %>% exp,
  or_uci95 = (presso.main$`Causal Estimate` + 1.96*presso.main$Sd) %>% exp
)
```

```{r}
res.aldh <- res_single %>% filter(SNP == "rs1229984")
res.aldh.or <- generate_odds_ratios(res.aldh)
aldh.res <- tibble(
  outcome = res.or$outcome[1],
  exposure = paste0(res.or$exposure[1], ", ", res.aldh.or$SNP[1], " - ALDH only"),
  method = "ALDH SNP only",
  nsnp = 1,
  pval = res.aldh.or$p[1],
  or = res.aldh.or$or[1],
  or_lci95 = res.aldh.or$or_lci95[1],
  or_uci95 = res.aldh.or$or_uci95[1]
)

res.outlier <- res_single %>% filter(SNP == "rs1229984")
res.outlier.or <- generate_odds_ratios(res.outlier)
outlier.res <- tibble(
  outcome = res.or$outcome[1],
  exposure = paste0(res.or$exposure[1], ", ", res.outlier.or$SNP[1], " - outlier only"),
  method = "outlier SNP only",
  nsnp = 1,
  pval = res.outlier.or$p[1],
  or = res.outlier.or$or[1],
  or_lci95 = res.outlier.or$or_lci95[1],
  or_uci95 = res.outlier.or$or_uci95[1]
)
```

```{r}
x <- Tryx$new(dat)
x$mrtryx()
```

```{r}
x$tryx.sig()
```

```{r}
x$analyse()
```

```{r}
#id_remove = c("id:ebi-a-GCST006572")
id_remove = NULL

x$analyse.mv(id_remove = id_remove)
```

```{r}
x$output$analyse.mv$estimates
```

```{r}
x$output$analyse.mv$plot
```

```{r}
x$output$analyse.mv
```

```{r}
#tryx.network(x$output)
```

```{r}
x$output$analyse.mv$estimates
```

```{r}
res.outlier <- res_single %>% filter(SNP == "rs1229984")
res.outlier.or <- generate_odds_ratios(res.outlier)
outlier.res <- tibble(
  outcome = res.or$outcome[1],
  exposure = paste0(res.or$exposure[1], ", ", res.outlier.or$SNP[1], " - outlier only"),
  method = "outlier SNP only",
  nsnp = 1,
  pval = res.outlier.or$p[1],
  or = res.outlier.or$or[1],
  or_lci95 = res.outlier.or$or_lci95[1],
  or_uci95 = res.outlier.or$or_uci95[1]
)
```

```{r}
res.or.liu2019.no23andme <- 
  res.or %>%
  select(outcome, exposure, method, nsnp, pval, or, or_lci95, or_uci95) %>%
  rbind(mrpresso.res, aldh.res, outlier.res) %>%
  rename_with(str_to_title) %>%
  rename("Number of SNPs" = Nsnp , "p" = Pval , "OR" = Or, "Lower 95% CI" = Or_lci95, "Upper 95% CI" = Or_uci95) %>%
  mutate(Outcome = "Alzheimer's Disease, Kunkle et al., 2019", Exposure = "Drinks Per Week, Liu et al., 2019, no 23andMe")
```

# {-}

```{r, eval = F}
write.csv(rbind(
  res.or.liu2019.full,
  res.or.liu2019.no23andme
  ), file = here("results", "alcohol_LOAD_model_results.csv"))
```

## Raw GSCAN summary statistics

"Evaluating the relationship between alcohol consumption, tobacco use, and cardiovascular disease: A multivariable Mendelian randomization study" exposure instrument list from the same study source came up with 71 SNPs. Possibly updated? Possibly using proxy SNPs? No looks like the paper identifies these 71 SNPs before any outcome ?

Full summary data from GSCAN below (https://conservancy.umn.edu/handle/11299/201564, with 23andMe removed, not available for public use). Should be identical to that presented in the paper, excluding 23andMe. Previous betas are reported based on the Wald statistic. Given the different ways of how each trait has been measured (binned, normalized, etc), we have since changed beta to be based on the chi square statistic instead using  the following transformation (below "beta", "statistic", and "af" refer to the respective column label in the summary statistic files):

For continuous traits with assumed standard deviation of 1 for the phenotype:
beta.est <- sign(beta.ori)*sqrt(statistic)*sqrt(1)/sqrt(2*N*af*(1-af));
beta.sd <- sqrt(1/(2*N*af*(1-af)))

For binary traits (var.y is the prevalence of cases. For our analysis, we took the mean prevalence across all studies in the analysis):
beta.est <- sign(beta.ori)*sqrt(statistic)/sqrt(2*N*af*(1-af))/sqrt(var.y));
beta.sd <- sqrt(1/sqrt(2*N*af*(1-af)*var.y));

```{r import_GSCAN_results, include = F, eval = F}
gscan.alc <- read.table(here("data", "DrinksPerWeek.txt"), header = T)

gscan.alc %>%
  group_by(RSID) %>%
  tally

#saveRDS(gscan.alc, here("data", "gscan_DrinksPerWeek.rda"))
#gscan.alc <- readRDS(here("data", "gscan_DrinksPerWeek.rda"))
```

```{r, include = F, eval = F}
exp <- 
  gscan.pfilter %>%
  rename_with(tolower) %>%
  rename(
    SNP = rsid,
    effect_allele = alt,
    other_allele = ref,
    eaf = af,
    pval = pvalue,
    samplesize = effective_n
  ) %>%
  filter(SNP != ".")

dat.exp <- format_data(exp)
dat.exp.clump <- clump_data(dat.exp, clump_p1 = 5e-8, clump_p2 = 5e-8)
```

```{r pfilter_GSCAN_results, include = F, eval = F}
#alc.exp.dat <- read_exposure_data(here("data", "gscan_DrinksPerWeek.rda"))
gscan.pfilter <- gscan.alc %>% filter(PVALUE < 5e-8)
gscan.pfilter %>% filter(RSID == ".")
```