---
title: "2012 Wave of HRS 1-sample Mendelian Randomization Assumption Testing"
author: "Kyle Abraham Campbell"
date: '2022-09-17'
output: html_document
---

```{r setup, include=FALSE}
library(boot)
library(formulaic)
library(ivtools)
library(knitr)
library(gtsummary) 
library(here)
library(parameters)
library(MendelianRandomization)
library(readxl)
library(skimr)
library(tidymodels)
library(tidyverse)
```

```{r setup2, eval=FALSE, include=FALSE}
#install.packages("devtools", type = "win.binary")
library(devtools)
library(remotes)
#remotes::install_github("qingyuanzhao/mr.raps")
library(mr.raps)
#devtools::install_github("WSpiller/RadialMR")
library(RadialMR)
#devtools::install_github("explodecomputer/tryx")
library(tryx)
#remotes::install_github("MRCIEU/TwoSampleMR")
#library(TwoSampleMR)
#install_github("tye27/mr.genius")
library(mr.genius)

knitr::opts_chunk$set(echo = TRUE)
```

as.data.frame is ESSENTIAL for ivglm to accept these as valid data inputs
```{r}
dat <- readRDS(here("data", "analytic", "2022-09-25.current.consumption.gscan.dpw.1smr.xs.rda")) %>% as_tibble(.name_repair = "universal") %>% mutate(DEGREE = factor(DEGREE))

#dat.ea <- readRDS(here("data", "analytic", "2022-08-23.gscan.dpw.1smr.xs.rda")) %>% filter(RACE == "White/Caucasian") %>%  as.data.frame()
#dat.aa <- readRDS(here("data", "analytic", "2022-08-23.gscan.dpw.1smr.xs.rda")) %>% filter(RACE == "Black or African American") %>% as.data.frame()
```

Get SNP IDs
```{r}
variants <-
  dat %>%
  # This approach assumes the SNP name still includes :'s
  #dplyr::select(matches(":\\D:\\D")) %>%
  # This approach uses updated repaired names
  dplyr::select(matches("\\.\\D\\.\\D")) %>%
  colnames()
```

```{r}
univariates = c("HHIDPN", "NAGE", "GENDER", "RACE", "DEGREE", "AD12", "drinkpweek", "logdrinkpweek",
                "R11DRINK", "R11DRINKD", "R11DRINKN", "APOE012",
                "R11SMOKEV", "NMARST", "R11CESD", "R11CONDE", "R11SAYRET",
                "PGS4_DPW_GSCAN19", "pgs.norm", "AD12.CIND.outcome", "AD12.Dementia.outcome", "AD12.binary") 
combined.pgs.covariates <- c("AncestryPC_1_5A", "AncestryPC_1_5B", "AncestryPC_1_5C", "AncestryPC_1_5D", "AncestryPC_1_5E")

model.vars <- c(univariates, combined.pgs.covariates, variants) %>% unique()
genetics <- c("PGS4_DPW_GSCAN19", "pgs.norm", variants)
```

The different kinds of models we build beyond naive and combined.pgs.covariates Ancestry-adjusted 5 PCs
```{r}
demographics <- c("NAGE", "GENDER", "DEGREE")
additional_covars <- c( "R11SMOKEV", "NMARST", "R11CESD", "R11CONDE", "R11SAYRET")
```

```{r}
dat <-
  dat %>%
  # Subset to only the relevant columns
  #dplyr::select(all_of(model.vars)) %>%
  mutate(AD12.binary = factor(AD12.binary))
```

Option to convert to long format with SNP as a single column
```{r, eval = F, include = F}
dat.long <-
  dat %>%
  pivot_longer(all_of(c(variants, "PGS4_DPW_GSCAN19", "pgs.norm")), names_to = "SNP", values_to = "Dosage")
```

```{r}
dat.ea <- dat %>% filter(RACE == "Non-Hispanic White, European ancestry") %>% as.data.frame()
dat.aa <- dat %>% filter(RACE == "Non-Hispanic Black, African ancestry") %>% as.data.frame()
```

# Relevance Assumption 

## Set models and helper functions
Source code for relevance model fitting
```{r}
source(here("scripts", "fit_relevance_models.R"))
```

Set linear regression tidy model
```{r}
lm_mod <-
  linear_reg()
```

Set logistic regression tidy model
```{r}
lr_mod <- 
  logistic_reg() %>% 
  set_engine("glm")
```

## European Ancestry
### Drinks per week exposure
Fit all relevance models for naive, ancestryPCs, demographic, and full models
```{r}
res_relevance_ea <- fit_relevance_models(instruments = genetics, exposure = "drinkpweek", tidy_model = lm_mod, data = dat.ea)
```

Display and plot results
```{r}
#TODO fix bug that's making Demo and Demo + APOE012 the same
ggplot(data = res_relevance_ea,
       mapping = aes(x = graph_F, color = Model, fill = Model)) +
  geom_histogram() +
  xlim(0, 20) +
  facet_wrap(facets = vars(Model)) +
  ggtitle(ggtitle(paste0("European ancestry histogram of instrument F-statistics, n = ", dim(dat.ea)[1])))

print(res_relevance_ea %>% filter(Instrument == "PGS4_DPW_GSCAN19"))

print(res_relevance_ea %>% filter(Instrument == "pgs.norm"))

ggplot(data = dat.ea, mapping = aes(x = PGS4_DPW_GSCAN19, y = drinkpweek)) + geom_point() + geom_smooth(method='lm', formula= y~x) +
  stat_cor() +
  ggtitle(paste0("European ancestry regress DPW on PGS4, n = ", dim(dat.ea)[1]))
```

#### Hypothesis: if PGSCAN GWAS excluded never drinkers or who quit before study measurement were set to missing -> better performance in non-never drinkers

```{r}
dat.ea %>% filter(everdrank == T) %>% dim
```

```{r}
table(dat.ea$R11DRINK, dat.ea$AD12)
table(dat.ea$R11DRINK, dat.ea$AD12) %>% chisq.test()
```

```{r}
table(dat.ea$everdrank, dat.ea$AD12)
table(dat.ea$everdrank, dat.ea$AD12) %>% chisq.test()
```

```{r}
res_relevance_ea_drinkers <- fit_relevance_models(instruments = genetics, exposure = "drinkpweek", tidy_model = lm_mod, data = dat.ea %>% filter(everdrank == T))
```

Display and plot results
```{r}
#TODO fix bug that's making Demo and Demo + APOE012 the same
ggplot(data = res_relevance_ea_drinkers,
       mapping = aes(x = graph_F, color = Model, fill = Model)) +
  geom_histogram() +
  xlim(0, 20) +
  facet_wrap(facets = vars(Model)) +
  ggtitle(ggtitle(paste0("European ancestry drinkers histogram of instrument F-statistics, n = ", dim(dat.ea)[1])))

print(res_relevance_ea_drinkers %>% filter(Instrument == "PGS4_DPW_GSCAN19"))

print(res_relevance_ea_drinkers %>% filter(Instrument == "pgs.norm"))

ggplot(data = dat.ea %>% filter(everdrank == T), mapping = aes(x = PGS4_DPW_GSCAN19, y = drinkpweek)) + geom_point() + geom_smooth(method='lm', formula= y~x) +
  stat_cor() +
  ggtitle(paste0("European ancestry regress DPW on PGS4, n = ", dim(dat.ea %>% filter(everdrank == T))[1]))
```

#### Hypothesis: if PGSCAN GWAS excluded not current drinkers or who quit before study measurement were set to missing -> better performance in not current drinkers

```{r}
dat.ea %>% filter(R11DRINK == 1) %>% dim
```

```{r}
table(dat.ea$R11DRINK, dat.ea$AD12)
table(dat.ea$R11DRINK, dat.ea$AD12) %>% chisq.test()
```

```{r}
table(dat.ea$everdrank, dat.ea$AD12)
table(dat.ea$everdrank, dat.ea$AD12) %>% chisq.test()
```

```{r}
res_relevance_ea_drinkers <- fit_relevance_models(instruments = genetics, exposure = "drinkpweek", tidy_model = lm_mod, data = dat.ea %>% filter(R11DRINK == 1))
```

Display and plot results
```{r}
#TODO fix bug that's making Demo and Demo + APOE012 the same
ggplot(data = res_relevance_ea_drinkers,
       mapping = aes(x = graph_F, color = Model, fill = Model)) +
  geom_histogram() +
  xlim(0, 20) +
  facet_wrap(facets = vars(Model)) +
  ggtitle(ggtitle(paste0("European ancestry current drinkers histogram of instrument F-statistics, n = ", dim(dat.ea)[1])))

print(res_relevance_ea_drinkers %>% filter(Instrument == "PGS4_DPW_GSCAN19"))

print(res_relevance_ea_drinkers %>% filter(Instrument == "pgs.norm"))

ggplot(data = dat.ea %>% filter(R11DRINK == 1), mapping = aes(x = PGS4_DPW_GSCAN19, y = drinkpweek)) + geom_point() + geom_smooth(method='lm', formula= y~x) +
  stat_cor() +
  ggtitle(paste0("European ancestry regress DPW on PGS4, n = ", dim(dat.ea %>% filter(R11DRINK == 1))[1]))
```

#### W/o 4 DPW outliers
Fit all relevance models for naive, ancestryPCs, demographic, and full models
```{r}
res_relevance_ea_dpw_lt_40_dpw_lt_40 <- fit_relevance_models(instruments = genetics, exposure = "drinkpweek", tidy_model = lm_mod, data = dat.ea %>% filter(drinkpweek < 40))
```

Display and plot results
```{r}
ggplot(data = res_relevance_ea_dpw_lt_40_dpw_lt_40,
       mapping = aes(x = graph_F, color = Model, fill = Model)) +
  geom_histogram() +
  xlim(0, 20) +
  facet_wrap(facets = vars(Model)) +
  ggtitle(ggtitle(paste0("<40 DPW European ancestry histogram of instrument F-statistics, n = ", dim(dat.ea %>% filter(drinkpweek < 40))[1])))

print(res_relevance_ea_dpw_lt_40_dpw_lt_40 %>% filter(Instrument == "PGS4_DPW_GSCAN19"))

print(res_relevance_ea_dpw_lt_40_dpw_lt_40 %>% filter(Instrument == "pgs.norm"))

ggplot(data = dat.ea %>% filter(drinkpweek < 40), mapping = aes(x = PGS4_DPW_GSCAN19, y = drinkpweek)) + geom_point() + geom_smooth(method='lm', formula= y~x) +
  ggtitle(paste0("European ancestry regress DPW on PGS4, n = ", dim(dat.ea %>% filter(drinkpweek < 40))[1]))
```

### Log-transformed drinks per week exposure
Fit all relevance models for naive, ancestryPCs, demographic, and full models
```{r}
res_relevance_ea_log <- fit_relevance_models(instruments = genetics, exposure = "drinkpweek", tidy_model = lm_mod, data = dat.ea)
```

Display and plot results
```{r}
ggplot(data = res_relevance_ea_log,
       mapping = aes(x = graph_F, color = Model, fill = Model)) +
  geom_histogram() +
  xlim(0, 20) +
  facet_wrap(facets = vars(Model)) +
  ggtitle(ggtitle(paste0("European ancestry histogram of instrument F-statistics, n = ", dim(dat.ea)[1])))

print(res_relevance_ea_log %>% filter(Instrument == "PGS4_DPW_GSCAN19"))

print(res_relevance_ea_log %>% filter(Instrument == "pgs.norm"))

ggplot(data = dat.ea, mapping = aes(x = PGS4_DPW_GSCAN19, y = logdrinkpweek)) + geom_point() + geom_smooth(method='lm', formula= y~x) +
  stat_cor() +
  ggtitle(paste0("European ancestry regress log-transformed DPW on PGS4, n = ", dim(dat.ea)[1]))
```
## African ancestry population - relevance
### Drinks per week exposure

Fit all relevance models for naive, ancestryPCs, demographic, and full models
```{r}
res_relevance_aa <- fit_relevance_models(instruments = genetics, exposure = "drinkpweek", tidy_model = lm_mod, data = dat.aa)
```

Display and plot results
```{r}
ggplot(data = res_relevance_aa,
       mapping = aes(x = graph_F, color = Model, fill = Model)) +
  geom_histogram() +
  xlim(0, 20) +
  facet_wrap(facets = vars(Model)) +
  ggtitle(ggtitle(paste0("African ancestry histogram of instrument F-statistics, n = ", dim(dat.aa)[1])))

print(res_relevance_aa %>% filter(Instrument == "PGS4_DPW_GSCAN19"))

print(res_relevance_aa %>% filter(Instrument == "pgs.norm"))

ggplot(data = dat.aa, mapping = aes(x = PGS4_DPW_GSCAN19, y = drinkpweek)) + geom_point() + geom_smooth(method='lm', formula= y~x) +
  stat_cor() +
  ggtitle(paste0("African ancestry regress DPW on PGS4, n = ", dim(dat.aa)[1]))
```

### Log-transformed drinkpweek

Fit all relevance models for naive, ancestryPCs, demographic, and full models
```{r}
res_relevance_aa_log <- fit_relevance_models(instruments = genetics, exposure = "logdrinkpweek", tidy_model = lm_mod, data = dat.aa)
```

Display and plot results
```{r}
ggplot(data = res_relevance_aa_log,
       mapping = aes(x = graph_F, color = Model, fill = Model)) +
  geom_histogram() +
  xlim(0, 20) +
  facet_wrap(facets = vars(Model)) +
  ggtitle(ggtitle(paste0("African ancestry histogram of instrument F-statistics, n = ", dim(dat.aa)[1])))

print(res_relevance_aa_log %>% filter(Instrument == "PGS4_DPW_GSCAN19"))

print(res_relevance_aa_log %>% filter(Instrument == "pgs.norm"))

ggplot(data = dat.aa, mapping = aes(x = PGS4_DPW_GSCAN19, y = logdrinkpweek)) + geom_point() + geom_smooth(method='lm', formula= y~x) +
  stat_cor() +
  ggtitle(paste0("African ancestry regress log-transformed DPW on PGS4, n = ", dim(dat.aa)[1]))
```

# Exclusion

```{r}
#TODO add African ancestry
tidy_parsnip_map_lr <- function(fit, map.x) {
  
  fit_res <- map2(fit, genetics, ~ tidy(.x, conf.int = TRUE, exponentiate = TRUE) %>% filter(term == .y))
  names(fit_res) <- genetics
  fit_res_df <- bind_rows(fit_res, .id = "Instrument")
  
  fit_glance <- map(fit, ~ glance(.x))
  names(fit_glance) <- genetics
  fit_glance_df <- bind_rows(fit_glance, .id = "Instrument") #%>% rename("F.stat" = statistic, "F.p.value" = p.value)
  
  fit_all <- left_join(fit_res_df, fit_glance_df, by = "Instrument")
  
  return(fit_all)
}
```

## European ancestry - Dementia
Warning: glm.fit: fitted probabilities numerically 0 or 1 occurred, 22 times
```{r}
fit <-
  map(
    .x = genetics,
    .f = ~ lr_mod %>%
      fit(formula(paste0("AD12.Dementia.outcome ~ ", .x)),
          data = dat.ea)
  )

# If you want to check out the observations that have probabilities near 0 or 1
#fit.GSCAN <- lr_mod %>%
#  fit(formula("AD12.Dementia.outcome ~ PGS4_DPW_GSCAN19"),
#      data = dat.ea)
#View(predict(fit.GSCAN, dat.ea, type = "prob"))

res_dementia_ea <- tidy_parsnip_map_lr(fit) %>%
  mutate(Ancestry = "European Ancestry")
```

## European ancestry - CIND
```{r}
fit <-
  map(
    .x = genetics,
    .f = ~ lr_mod %>%
      fit(formula(paste0("AD12.CIND.outcome ~ ", .x)),
          data = dat.ea)
  )

res_cind_ea <- tidy_parsnip_map_lr(fit) %>%
  mutate(Ancestry = "European Ancestry")
```

## Fit covariate -> instrument
```{r}
independent.vars <- c("NAGE", "GENDER", "DEGREE", "APOE012", "R11SMOKEV", "NMARST", "R11CESD", "R11CONDE", "R11SAYRET", "AncestryPC_1_5A", "AncestryPC_1_5B", "AncestryPC_1_5C", "AncestryPC_1_5D", "AncestryPC_1_5E")
```

```{r}
genetics.enum <- rep(genetics, times = length(independent.vars))
independent.vars.enum <- rep(independent.vars, times = length(genetics))
```

```{r}
# Fit the linear model across all instrument and covariate combinations for F-test
fit <-
  map2(
    .x = genetics.enum,
    .y = independent.vars.enum,
    .f = ~ lm_mod %>%
      fit(formula(paste0(.x, " ~ ", .y)),
          data = dat.ea)
  )

#fit_res <- map2(fit, genetics, ~ tidy(.x) %>% filter(term == .y))
#names(fit_res) <- genetics
#fit_res_df <- bind_rows(fit_res, .id = "Instrument")

# Collect the results into a data frame
# Add FDR-adjusted p-value
fit_glance <- map(fit, ~ glance(.x))
names(fit_glance) <- paste0(genetics.enum, "+", independent.vars.enum)
fit_glance_df <-
  bind_rows(fit_glance, .id = "Instrument") %>%
  rename("F.stat" = statistic, "F.p.value" = p.value) %>%
  mutate(F.p.value.fdr = p.adjust(F.p.value, method = "fdr")) %>%
  rowwise() %>%
  mutate(predictor = strsplit(Instrument,"+", fixed = T)[[1]][2]) %>%
  mutate(Instrument = strsplit(Instrument,"+", fixed = T)[[1]][1]) %>%
  dplyr::select(Instrument, predictor, everything())

#fit_all <- left_join(fit_res_df, fit_glance_df, by = "Instrument")
```

```{r}
fit_glance_df %>% group_by(Instrument) %>% filter(F.p.value < 0.05) %>% tally()
fit_glance_df %>% group_by(Instrument) %>% filter(F.p.value.fdr < 0.05) %>% tally()
fit_glance_df %>% group_by(Instrument) %>% filter(F.p.value < 0.05) %>% tally() %>% pull(n) %>% hist
fit_glance_df %>% group_by(Instrument) %>% filter(F.p.value.fdr < 0.05) %>% tally() %>% pull(n) %>% hist
```

```{r}
fit_glance_df %>% group_by(predictor) %>% filter(F.p.value < 0.05) %>% tally()
fit_glance_df %>% group_by(predictor) %>% filter(F.p.value.fdr < 0.05) %>% tally()
```

Most associations were for the ancestry PCs
```{r}
#View(fit_glance_df %>% filter(F.p.value < 0.05))
#View(fit_glance_df %>% filter(F.p.value.fdr < 0.05))
```
