---
title: "Mendelian Randomization tests of instrument strength, independence, and exlcusion restriction assumptions"
author: "Kyle Abraham Campbell"
date: "4/25/2022"
output: html_document
---

```{r setup, include=FALSE}
library(boot)
library(formulaic)
library(ivtools)
library(knitr)
library(gtsummary) 
library(here)
library(parameters)
library(MendelianRandomization)
library(readxl)
library(skimr)
library(tidymodels)
library(tidyverse)

source(here("scripts", "gscan_dpw_1smr_xs_model_functions.R"))
```

```{r setup2, eval=FALSE, include=FALSE}
#install.packages("devtools", type = "win.binary")
library(devtools)
library(remotes)
#remotes::install_github("qingyuanzhao/mr.raps")
library(mr.raps)
#devtools::install_github("WSpiller/RadialMR")
library(RadialMR)
#devtools::install_github("explodecomputer/tryx")
library(tryx)
#remotes::install_github("MRCIEU/TwoSampleMR")
#library(TwoSampleMR)
#install_github("tye27/mr.genius")
library(mr.genius)

knitr::opts_chunk$set(echo = TRUE)
```

```{r}
#write("TMPDIR = 'C:\r_tmpdir'", file=file.path(Sys.getenv('R_USER'), '.Renviron'))
```

as.data.frame is ESSENTIAL for ivglm to accept these as valid data inputs
```{r}
#TODO why is this formatting here?
dat <- readRDS(here("data", "analytic", "2022-09-25.current.consumption.gscan.dpw.1smr.xs.rda")) %>% as_tibble(.name_repair = "universal") %>% mutate(DEGREE = factor(DEGREE))
```

Get SNP IDs
```{r}
variants <-
  dat %>%
  # This approach assumes the SNP name still includes :'s
  #dplyr::select(matches(":\\D:\\D")) %>%
  # This approach uses updated repaired names
  dplyr::select(matches("\\.\\D\\.\\D")) %>%
  colnames()
```

```{r}
univariates = c("HHIDPN", "NAGE", "GENDER", "RACE", "DEGREE", "AD12", "drinkpweek", "logdrinkpweek",
                "R11DRINK", "R11DRINKD", "R11DRINKN", "APOE012",
                "R11SMOKEV", "NMARST", "R11CESD", "R11CONDE", "R11SAYRET",
                "PGS4_DPW_GSCAN19", "pgs.norm", "AD12.CIND.outcome", "AD12.Dementia.outcome", "AD12.binary") 
combined.pgs.covariates <- c("AncestryPC_1_5A", "AncestryPC_1_5B", "AncestryPC_1_5C", "AncestryPC_1_5D", "AncestryPC_1_5E")

model.vars <- c(univariates, combined.pgs.covariates, variants) %>% unique()
genetics <- c("PGS4_DPW_GSCAN19", "pgs.norm", variants)
```

The different kinds of models we build beyond naive and combined.pgs.covariates Ancestry-adjusted 5 PCs
```{r}
demographics <- c("NAGE", "GENDER", "DEGREE")
additional_covars <- c( "R11SMOKEV", "NMARST", "R11CESD", "R11CONDE", "R11SAYRET")
```

```{r}
dat %>% filter(R11DRINK == "Drinks") %>% pull(NAGE) %>% mean()
dat %>% filter(R11DRINK == "Drinks") %>% pull(drinkpweek) %>% mean()
dat %>% filter(R11DRINK == "Drinks") %>% pull(drinkpweek) %>% sd()
```

```{r}
dat <-
  dat %>%
  dplyr::select(all_of(model.vars)) %>%
  mutate(AD12.binary = factor(AD12.binary))
```

Option to convert to long format with SNP as a single column
```{r, eval = F, include = F}
dat.long <-
  dat %>%
  pivot_longer(all_of(c(variants, "PGS4_DPW_GSCAN19", "pgs.norm")), names_to = "SNP", values_to = "Dosage")
```

```{r}
dat.ea <- dat %>% filter(RACE == "Non-Hispanic White, European ancestry") %>% as.data.frame()
dat.aa <- dat %>% filter(RACE == "Non-Hispanic Black, African ancestry" ) %>% as.data.frame()
```

# MendelianRandomization R package for IV estimation and robust senstivity analyses of pleiotropy

## Pleiotropy

### Fit Instrument -> Exposure model in EA

```{r}
# Drop the PGS scores for the individual SNP analysis
snps <- genetics[-seq(1,2)]
```

```{r, eval = F}
# Fit the linear model across all instruments, adjusted for ancestry covars
fit <-
  map(
    .x = snps,
    .f = ~ lm_mod %>%
      fit(formula(paste0("drinkpweek ~ ", .x, " + ", paste(c(additional_covars, demographics, combined.pgs.covariates), collapse = " + "))),
          data = dat.ea)
  )

fit_res <- map2(fit, snps, ~ tidy(.x) %>% filter(term == .y))
names(fit_res) <- snps
fit_res_df <- bind_rows(fit_res, .id = "Instrument")

fit_glance <- map(fit, ~ glance(.x))
names(fit_glance) <- snps
fit_glance_df <- bind_rows(fit_glance, .id = "Instrument") %>% rename("F.stat" = statistic, "F.p.value" = p.value)

fit_all <- left_join(fit_res_df, fit_glance_df, by = "Instrument")

bx <- fit_all$estimate
bxse <- fit_all$std.error
```

### Fit Instrument -> Outcome CIND model in EA
```{r, eval = F}
fit <-
  map(
    .x = snps,
    .f = ~ lr_mod %>%
      fit(formula(paste0("AD12.CIND.outcome ~ ", .x, " + ", paste(c(additional_covars, demographics, combined.pgs.covariates), collapse = " + "))),
          data = dat.ea)
  )

fit_res <- map2(fit, snps, ~ tidy(.x, conf.int = TRUE, exponentiate = FALSE) %>% filter(term == .y))
names(fit_res) <- snps
fit_res_df <- bind_rows(fit_res, .id = "Instrument")

fit_glance <- map(fit, ~ glance(.x))
names(fit_glance) <- snps
fit_glance_df <- bind_rows(fit_glance, .id = "Instrument") #%>% rename("F.stat" = statistic, "F.p.value" = p.value)

fit_all <- left_join(fit_res_df, fit_glance_df, by = "Instrument")

res.cind.ea <- fit_all %>%
  mutate(Ancestry = "European Ancestry")

by <- fit_all$estimate
byse <- fit_all$std.error

MRInputObject <- 
  mr_input(bx = bx,
           bxse = bxse,
           by = by,
           byse = byse)

MRAllObject_all <- mr_allmethods(MRInputObject, method = "all")

MRAllObject_all

mr_plot(MRInputObject, error = TRUE, orientate = FALSE, line = "ivw")

mr_plot(MRInputObject, error = TRUE, orientate = FALSE, line = "ivw", interactive = F)
```

```{r Naive models, eval = F}
mr.cind.ea <- run_snp_mr_naive_model(exposure = "logdrinkpweek", outcome = "AD12.CIND.outcome", dat = dat.ea)
mr.dem.ea <- run_snp_mr_naive_model(exposure = "logdrinkpweek", outcome = "AD12.Dementia.outcome", dat = dat.ea)
mr.cind.aa <- run_snp_mr_naive_model(exposure = "logdrinkpweek", outcome = "AD12.CIND.outcome", dat = dat.aa)
mr.dem.aa <- run_snp_mr_naive_model(exposure = "logdrinkpweek", outcome = "AD12.Dementia.outcome", dat = dat.aa)
mr.binary.ea <- run_snp_mr_naive_model(exposure = "logdrinkpweek", outcome = "AD12.binary", dat = dat.ea)
mr.binary.aa <- run_snp_mr_naive_model(exposure = "logdrinkpweek", outcome = "AD12.binary", dat = dat.aa)
```

```{r}
#naive.models <- readRDS(here("results", "analytic", paste0("one_smr_pcs_2022-08-23.rda")))

naive.models <- list(mr.cind.ea = mr.cind.ea, mr.dem.ea = mr.dem.ea, mr.cind.aa = mr.cind.aa, mr.dem.aa = mr.dem.aa, mr.binary.ea = mr.binary.ea, mr.binary.aa = mr.binary.aa)

#saveRDS(naive.models, here("results", "analytic", "current_consumers", paste0("one_smr_naive_", Sys.Date(), ".rda")))
```

```{r Ancestry PCs models, eval = F}
mr.cind.ea <- run_snp_mr(exposure = "logdrinkpweek", outcome = "AD12.CIND.outcome", covariates = combined.pgs.covariates, dat = dat.ea)
mr.dem.ea <- run_snp_mr(exposure = "logdrinkpweek", outcome = "AD12.Dementia.outcome", covariates = combined.pgs.covariates, dat = dat.ea)
mr.cind.aa <- run_snp_mr(exposure = "logdrinkpweek", outcome = "AD12.CIND.outcome", covariates = combined.pgs.covariates, dat = dat.aa)
mr.dem.aa <- run_snp_mr(exposure = "logdrinkpweek", outcome = "AD12.Dementia.outcome", covariates = combined.pgs.covariates, dat = dat.aa)
mr.binary.ea <- run_snp_mr(exposure = "logdrinkpweek", outcome = "AD12.binary", covariates = combined.pgs.covariates, dat = dat.ea)
mr.binary.aa <- run_snp_mr(exposure = "logdrinkpweek", outcome = "AD12.binary", covariates = combined.pgs.covariates, dat = dat.aa)
```

```{r}
ancestry.models <- readRDS(here("results", "analytic", paste0("one_smr_pcs_2022-08-23.rda")))

#ancestry.models <- list(mr.cind.ea = mr.cind.ea, mr.dem.ea = mr.dem.ea,
#               mr.cind.aa = mr.cind.aa, mr.dem.aa = mr.dem.aa,
#               mr.binary.ea = mr.binary.ea, mr.binary.aa = mr.binary.aa)

#saveRDS(ancestry.models, here("results", "analytic", paste0("one_smr_pcs_", Sys.Date(), ".rda")))
```

```{r}
map(ancestry.models, ~ mr_plot(.x$MRAllObject))
```

```{r Add demographics to ancestry PCs model, eval = F}
mr.cind.ea <- run_snp_mr(exposure = "logdrinkpweek", outcome = "AD12.CIND.outcome", covariates = c(demographics, combined.pgs.covariates), dat = dat.ea)
mr.dem.ea <- run_snp_mr(exposure = "logdrinkpweek", outcome = "AD12.Dementia.outcome", covariates = c(demographics, combined.pgs.covariates), dat = dat.ea)
mr.cind.aa <- run_snp_mr(exposure = "logdrinkpweek", outcome = "AD12.CIND.outcome", covariates = c(demographics, combined.pgs.covariates), dat = dat.aa)
mr.dem.aa <- run_snp_mr(exposure = "logdrinkpweek", outcome = "AD12.Dementia.outcome", covariates = c(demographics, combined.pgs.covariates), dat = dat.aa)
mr.binary.ea <- run_snp_mr(exposure = "logdrinkpweek", outcome = "AD12.binary", covariates = c(demographics, combined.pgs.covariates), dat = dat.ea)
mr.binary.aa <- run_snp_mr(exposure = "logdrinkpweek", outcome = "AD12.binary", covariates = c(demographics, combined.pgs.covariates), dat = dat.aa)
```

```{r, eval = F}
demographics.models <- list(mr.cind.ea = mr.cind.ea, mr.dem.ea = mr.dem.ea,
               mr.cind.aa = mr.cind.aa, mr.dem.aa = mr.dem.aa,
               mr.binary.ea = mr.binary.ea, mr.binary.aa = mr.binary.aa)
```

```{r Add APOE012 to demographic model, eval = F}
mr.cind.ea <- run_snp_mr(exposure = "logdrinkpweek", outcome = "AD12.CIND.outcome", covariates = c(demographics, "APOE012", combined.pgs.covariates), dat = dat.ea)
mr.dem.ea <- run_snp_mr(exposure = "logdrinkpweek", outcome = "AD12.Dementia.outcome", covariates = c(demographics, "APOE012", combined.pgs.covariates), dat = dat.ea)
mr.cind.aa <- run_snp_mr(exposure = "logdrinkpweek", outcome = "AD12.CIND.outcome", covariates = c(demographics, "APOE012", combined.pgs.covariates), dat = dat.aa)
mr.dem.aa <- run_snp_mr(exposure = "logdrinkpweek", outcome = "AD12.Dementia.outcome", covariates = c(demographics, "APOE012", combined.pgs.covariates), dat = dat.aa)
mr.binary.ea <- run_snp_mr(exposure = "logdrinkpweek", outcome = "AD12.binary", covariates = c(demographics, "APOE012", combined.pgs.covariates), dat = dat.ea)
mr.binary.aa <- run_snp_mr(exposure = "logdrinkpweek", outcome = "AD12.binary", covariates = c(demographics, "APOE012", combined.pgs.covariates), dat = dat.aa)
```

```{r, eval = F}
demographics.apoe.models <- list(mr.cind.ea = mr.cind.ea, mr.dem.ea = mr.dem.ea, mr.cind.aa = mr.cind.aa, mr.dem.aa = mr.dem.aa, mr.binary.ea = mr.binary.ea, mr.binary.aa = mr.binary.aa)
```

```{r}
demographics.apoe.models <- readRDS(here("results", "analytic", "current_consumers", paste0("one_smr_demographics_apoe_2022-09-25.rda")))

#saveRDS(demographics.apoe.models, here("results", "analytic", "current_consumers", paste0("one_smr_demographics_apoe_", Sys.Date(), ".rda")))
```

```{r Full Model, eval = F}
mr.cind.ea <- run_snp_mr(exposure = "logdrinkpweek", outcome = "AD12.CIND.outcome", covariates = c(demographics, additional_covars, combined.pgs.covariates), dat = dat.ea)
mr.dem.ea <- run_snp_mr(exposure = "logdrinkpweek", outcome = "AD12.Dementia.outcome", covariates = c(demographics, additional_covars, combined.pgs.covariates), dat = dat.ea)
mr.cind.aa <- run_snp_mr(exposure = "logdrinkpweek", outcome = "AD12.CIND.outcome", covariates = c(demographics, additional_covars, combined.pgs.covariates), dat = dat.aa)
mr.dem.aa <- run_snp_mr(exposure = "logdrinkpweek", outcome = "AD12.Dementia.outcome", covariates = c(demographics, additional_covars, combined.pgs.covariates), dat = dat.aa)
mr.binary.ea <- run_snp_mr(exposure = "logdrinkpweek", outcome = "AD12.binary", covariates = c(demographics, additional_covars, combined.pgs.covariates), dat = dat.ea)
mr.binary.aa <- run_snp_mr(exposure = "logdrinkpweek", outcome = "AD12.binary", covariates = c(demographics, additional_covars, combined.pgs.covariates), dat = dat.aa)
```

```{r, eval = F}
full.models <- list(mr.cind.ea = mr.cind.ea, mr.dem.ea = mr.dem.ea,
               mr.cind.aa = mr.cind.aa, mr.dem.aa = mr.dem.aa,
               mr.binary.ea = mr.binary.ea, mr.binary.aa = mr.binary.aa)
```

```{r}
full.models <- readRDS(here("results", "analytic", paste0("one_smr_full_2022-08-23.rda")))

#saveRDS(full.models, here("results", "analytic", paste0("one_smr_full_", Sys.Date(), ".rda")))
```

```{r}
mr.res <- demographics.apoe.models
```

```{r}
map(mr.res, ~ mr_plot(.x$MRAllObject))
```

For ea.dementia, SNP 60 is a high variance outlier
```{r}
map(mr.res, ~ mr_plot(.x$MRInputObject, interactive = T, error = T))
```

```{r}
map(mr.res, ~ mr_egger(.x$MRInputObject))
```

```{r}
input <- mr.res$mr.cind.ea$MRInputObject
res <- mr.res$mr.cind.ea$MRAllObject
```

```{r}
mr_plot(input, orientate = T, interactive = F) + xlim(c(-0.5, 0.5)) + ylim(c(-1, 1))
mr_plot(res, orientate = TRUE, line = "ivw", interactive = T) + xlim(c(-0.2, 0.2)) + ylim(c(-0.5, 0.5))
```

SNPs 76 and 60 are highly unstable
```{r}
res.snps <- res@Data

res.snps.df <- data.frame(
  snp = variants,
  betaX = res.snps@betaX,
  betaY = res.snps@betaY,
  betaXse = res.snps@betaXse,
  betaYse = res.snps@betaYse
) %>%
  mutate(
    betaXlci = betaX - 1.96*betaXse,
    betaXuci = betaX + 1.96*betaXse,
    betaYlci = betaY - 1.96*betaYse,
    betaYuci = betaY + 1.96*betaYse
    )

res.snps.df %>% filter(betaYse > 10) %>% pull(snp)
```

```{r}
mr_plot(input, error = T, orientate = TRUE, line = "MREgger")
```

Repeat but dropping 60 and 76
```{r}
snp.subset <- res.snps.df %>% filter(betaYse < 10)

bx = snp.subset$betaX
bxse = snp.subset$betaXse
by = snp.subset$betaY
byse = snp.subset$betaYse

MRInputObject <- 
  mr_input(bx = bx,
           bxse = bxse,
           by = by,
           byse = byse)

MRAllObject_all <- mr_allmethods(MRInputObject, method = "all")
```

```{r}
mr_plot(MRAllObject_all)
```

```{r}
show(MRAllObject_all) %>% View
```


## G-estimation

Custom function to summarize ivglm output
```{r}
# Function accepts a fitted ivglm model and outputs tidy-formatted summary tibble with exponentiated effect estimate and 95% CIs
tidy_ivglm_summary <- function(mod) {
  summary <- summary(mod)
  tibble(
    term = attr(summary$coefficients, "dimnames")[[1]],
    estimate = round((summary$coefficients[1]), 2),
    std.error = round(summary$coefficients[2], 2),
    statistic = round(summary$coefficients[3], 2),
    p.value = round(summary$coefficients[4], 2),
    conf.low = round((confint(mod)[[1]]), 2),
    conf.high = round((confint(mod)[[2]]), 2)
  )
}

tidy_ivglm_logit_summary <- function(mod) {
  summary <- summary(mod)
  tibble(
    term = attr(summary$coefficients, "dimnames")[[1]],
    estimate = round(exp(summary$coefficients[1]), 2),
    std.error = round(exp(summary$coefficients[2]), 2),
    statistic = round(summary$coefficients[3], 2),
    p.value = round(summary$coefficients[4], 2),
    conf.low = round(exp(confint(mod)[[1]]), 2),
    conf.high = round(exp(confint(mod)[[2]]), 2)
  )
}
```

### CIND vs. Normal, EA 
Provides unbiased OR estimates
```{r}
nrow(dat.ea)
summary(dat.ea$AD12.CIND.outcome)
```

### Dementia vs. Normal, EA

```{r}
nrow(dat.ea)
summary(dat.ea$AD12.Dementia.outcome)
```

```{r}
run_variant_dementia_ivglm <- function(data, variant, sample) {
  
  fitZ.L <- glm(formula = as.formula(paste0(add.backtick(variant), " ~ GENDER + NAGE + AncestryPC_1_5A + AncestryPC_1_5B + AncestryPC_1_5C + AncestryPC_1_5D + AncestryPC_1_5E")),
                data = data)
  summary(fitZ.L)
  
  fitY.LZX <-
    glm(
      formula = as.formula(paste0( "AD12.Dementia.outcome ~ ", add.backtick(variant), " + everdrinker + GENDER + NAGE +  AncestryPC_1_5A + AncestryPC_1_5B + AncestryPC_1_5C + AncestryPC_1_5D + AncestryPC_1_5E")),
      family = "binomial",
      data = data
    )
  summary(fitY.LZX)
  
  fitIV_g  <-
    ivglm(
      estmethod = "g",
      X = "everdrinker",
      Y = "AD12.Dementia.outcome",
      fitZ.L = fitZ.L,
      fitY.LZX = fitY.LZX,
      data = data,
      link = "logit"
    )
  
  res <- tidy_ivglm_summary(fitIV_g)
  res$outcome <- "Dementia"
  res$sample <- sample
  res$variant <- variant
  
  # Return fitIV_g or res
  return(fitIV_g)
}

run_variant_cind_ivglm <- function(data, variant, sample) {
  
  fitZ.L <- glm(formula = as.formula(paste0(add.backtick(variant), " ~ GENDER + NAGE +  AncestryPC_1_5A + AncestryPC_1_5B + AncestryPC_1_5C + AncestryPC_1_5D + AncestryPC_1_5E")),
                data = data)
  summary(fitZ.L)
  
  fitY.LZX <-
    glm(
      formula = as.formula(paste0( "AD12.CIND.outcome ~ ", add.backtick(variant), " + everdrinker + GENDER + NAGE +  AncestryPC_1_5A + AncestryPC_1_5B + AncestryPC_1_5C + AncestryPC_1_5D + AncestryPC_1_5E")),
      family = "binomial",
      data = data
    )
  summary(fitY.LZX)
  
  fitIV_g  <-
    ivglm(
      estmethod = "g",
      X = "everdrinker",
      Y = "AD12.CIND.outcome",
      fitZ.L = fitZ.L,
      fitY.LZX = fitY.LZX,
      data = data,
      link = "logit"
    )
  
  res <- tidy_ivglm_summary(fitIV_g)
  res$outcome <- "CIND"
  res$sample <- sample
  res$variant <- variant
  
  # Return fitIV_g or res
  return(fitIV_g)
}
# Example with a single variant
x = run_variant_dementia_ivglm(data = dat.ea, sample = "European Ancestry", variant = "9:109345993:T:C")
run_variant_cind_ivglm(data = dat.ea, sample = "European Ancestry", variant = "9:109345993:T:C")
run_variant_dementia_ivglm(data = dat.aa, sample = "Black or African American", variant = "9:109345993:T:C")
run_variant_cind_ivglm(data = dat.aa, sample = "Black or African American", variant = "9:109345993:T:C")
```

```{r}
res.dem.ea <- lapply(colnames(variants), function(x) run_variant_dementia_ivglm(data = dat.ea, variant = x, sample = "European Ancestry")) %>% bind_rows
res.dem.aa <- lapply(colnames(variants), function(x) run_variant_dementia_ivglm(data = dat.ea, variant = x, sample = "Black or African American")) %>% bind_rows
res.cind.ea <- lapply(colnames(variants), function(x) run_variant_cind_ivglm(data = dat.ea, variant = x, sample = "European Ancestry")) %>% bind_rows
res.cind.aa <- lapply(colnames(variants), function(x) run_variant_cind_ivglm(data = dat.ea, variant = x, sample = "Black or African American")) %>% bind_rows
```

```{r}
res <- rbind(res.dem.ea, res.dem.aa, res.cind.ea, res.cind.aa)
```

```{r}
res.graph <- 
  res %>%
  filter(conf.high < 10)

res %>%
  filter(conf.high > 10) %>%
  tally

forest <- ggplot(data=res.graph,
                 aes(
                   x=stringr::str_wrap(sample, 10),
                   y=estimate,
                   ymin=conf.low,
                   ymax=conf.high,
                   shape = outcome,
                   color = outcome)) +
  xlab("") +
  ylab("Odds Ratio for Ever Drinking") +
  labs(shape = "Outcome", color = "Outcome") +
  geom_hline(yintercept = 1, linetype = 2) +
  geom_pointrange(position=position_dodge(width = 1)) +
  coord_cartesian(ylim = c(0, 2.5)) +
  coord_flip() +
  theme_bw() +
  theme(text = element_text(size = 18))
forest
```

# Two-stage least squares

According to the R package documentation, for two-stage least squares, we use binomial links for the glm() call, and apply the ctrl=T argument in the ivglm() call without specifying a link function in the ivglm() call
```{r}
run_variant_dementia_ivglm <- function(data, variant, sample) {
  
  fitY.LX <-
    glm(
      formula = as.formula(paste0( "AD12.CIND.outcome ~ + everdrinker + GENDER + NAGE +  AncestryPC_1_5A + AncestryPC_1_5B + AncestryPC_1_5C + AncestryPC_1_5D + AncestryPC_1_5E")),
      family = "binomial",
      data = data
    )
  summary(fitY.LX)
  
  fitX.LZ <-
    glm(
      formula = as.formula(paste0( "everdrinker ~ ", add.backtick(variant), " + GENDER + NAGE +  AncestryPC_1_5A + AncestryPC_1_5B + AncestryPC_1_5C + AncestryPC_1_5D + AncestryPC_1_5E")),
      family = "binomial",
      data = data
    )
  summary(fitX.LZ)
  
  fitIV_g  <-
    ivglm(
      estmethod = "ts",
      X = "everdrinker",
      Y = "AD12.Dementia.outcome",
      fitX.LZ = fitX.LZ,
      fitY.LX = fitY.LX,
      data = data,
      ctrl = T
    )
  
  res <- tidy_ivglm_logit_summary(fitIV_g)
  res$outcome <- "Dementia"
  res$sample <- sample
  res$variant <- variant
  
  # Return fitIV_g or res
  return(fitIV_g)
}

run_variant_cind_ivglm <- function(data, variant, sample) {
  
  fitY.LX <-
    glm(
      formula = as.formula(paste0( "AD12.CIND.outcome ~ + everdrinker + GENDER + NAGE +  AncestryPC_1_5A + AncestryPC_1_5B + AncestryPC_1_5C + AncestryPC_1_5D + AncestryPC_1_5E")),
      family = "binomial",
      data = data
    )
  summary(fitY.LX)
  
  fitX.LZ <-
    glm(
      formula = as.formula(paste0( "everdrinker ~ ", add.backtick(variant), " + GENDER + NAGE +  AncestryPC_1_5A + AncestryPC_1_5B + AncestryPC_1_5C + AncestryPC_1_5D + AncestryPC_1_5E")),
      family = "binomial",
      data = data
    )
  summary(fitX.LZ)
  
  fitIV_g  <-
    ivglm(
      estmethod = "ts",
      X = "everdrinker",
      Y = "AD12.CIND.outcome",
      fitX.LZ = fitX.LZ,
      fitY.LX = fitY.LX,
      data = data,
      ctrl = T
    )
  
  res <- tidy_ivglm_logit_summary(fitIV_g)
  res$outcome <- "CIND"
  res$sample <- sample
  res$variant <- variant
  
  # Return fitIV_g or
  return(fitIV_g)
}

# Example with a single variant
x = run_variant_dementia_ivglm(data = dat.ea, sample = "European Ancestry", variant = "9:109345993:T:C")
summary(x)

run_variant_dementia_ivglm(data = dat.ea, sample = "European Ancestry", variant = "9:109345993:T:C") %>% summary
run_variant_cind_ivglm(data = dat.ea, sample = "European Ancestry", variant = "9:109345993:T:C") %>% summary
run_variant_dementia_ivglm(data = dat.aa, sample = "Black or African American", variant = "9:109345993:T:C") %>% summary
run_variant_cind_ivglm(data = dat.aa, sample = "Black or African American", variant = "9:109345993:T:C") %>% summary
```

```{r}
tidy_ivglm_logit_summary <- function(mod) {
  summary <- summary(mod)
  tibble(
    term = attr(summary$coefficients, "dimnames")[[1]][2],
    estimate = round(exp(summary$coefficients[1]), 2),
    std.error = round(exp(summary$coefficients[2]), 2),
    statistic = round(summary$coefficients[3], 2),
    p.value = round(summary$coefficients[4], 2),
    conf.low = round(exp(confint(mod)[[1]]), 2),
    conf.high = round(exp(confint(mod)[[2]]), 2)
  )
}
```

```{r}
tidy_ivglm_logit_summary(x)
mod <- x
summary <- summary(x)
# term
term <- attr(x.sum$coefficients, "dimnames")[[1]][2]

# p.value
p.value <- round(summary$coefficients[2, 4], 2)

# conf.low
conf.low <- confint(mod)[2,1]
conf.low
# conf.high
conf.high <- confint(mod)[2,2]
conf.high
```

```{r}
x.sum
```

Ever drinker models
```{r}
data = dat.ea
variant = "9:109345993:T:C"
outcome = "AD12.Dementia.outcome"
covariates = c("GENDER", "NAGE", "AncestryPC_1_5A", "AncestryPC_1_5B", "AncestryPC_1_5C", "AncestryPC_1_5D", "AncestryPC_1_5E")
precision.covariates = c("NAGE", "GENDER", "DEGREE", "drinkpweek",
                "APOE012", "R11SMOKEV", "NMARST", "R11CESD", "R11CONDE", "R11SAYRET")
ancestry.covars = c("AncestryPC_1_5A", "AncestryPC_1_5B", "AncestryPC_1_5C", "AncestryPC_1_5D", "AncestryPC_1_5E")


ivglm_model(data = dat.ea, ancestry = "ea", variant = "9:109345993:T:C", outcome = "AD12.Dementia.outcome", covariates = covariates)

# IVGLM model for both ts and g-est with everdrinker binary exposure, continuous IV, and binary outcome, implemented with ivtools package
ivglm_model <- function(data, ancestry, variant, outcome, covariates) {

  ## Two-stage estimation
  fitX.LZ <-
    glm(formula = as.formula(
      paste0(
        "everdrinker ~ ",
        add.backtick(variant),
        " + ", paste(covariates, collapse = " + ")
      )
    ),
    family = "binomial",
    data = data)
  fitY.LX <-
    glm(formula = as.formula(
      paste0(
        outcome,
        " ~ + everdrinker + ", paste(covariates, collapse = " + ")
      )
    ),
    family = "binomial",
    data = data)
  fitIV_ts  <-
    ivglm(
      estmethod = "ts",
      X = "everdrinker",
      Y = outcome,
      fitX.LZ = fitX.LZ,
      fitY.LX = fitY.LX,
      data = data,
      ctrl = T
    )
  
  # summarize ts model
  summary.ts <- summary(fitIV_ts)
    
  # export to results table
  res.ts <- data.frame(
    variant = variant,
    model = "ts",
    ancestry = ancestry,
    term = "everdrinker",
    estimate = round(exp(fitIV_ts$est["everdrinker"]), 2),
    std.error = round(summary.ts$coefficients[2,2], 2),
    statistic = round(summary.ts$coefficients[2,3], 2),
    p.value =  round(summary.ts$coefficients[2,4], 2),
    conf.low = round(exp(confint(fitIV_ts)[2,1]), 2),
    conf.high = round(exp(confint(fitIV_ts)[2,2]), 2),
    covariates = paste(covariates, collapse = " + ")
  )
  
  ## G-estimation
  fitZ.L <-
    glm(formula = as.formula(
      paste0(
        add.backtick(variant),
        " ~ ", paste(covariates, collapse = " + ")
      )
    ),
    data = data)
  
  fitY.LZX <-
    glm(formula = as.formula(
      paste0(
        outcome,
        " ~ ",
        add.backtick(variant),
        " + everdrinker + ", paste(covariates, collapse = " + ")
      )
    ),
    family = "binomial",
    data = data)
  
  fitIV_g  <-
    ivglm(
      estmethod = "g",
      X = "everdrinker",
      Y = outcome,
      fitZ.L = fitZ.L,
      fitY.LZX = fitY.LZX,
      data = data,
      link = "logit"
    )

  # Summarize g-est model
  summary.g <- summary(fitIV_g)
  
  # Export g-est model results
  res.g <- data.frame(
    variant = variant,
    model = "g",
    ancestry = ancestry,
    term = "everdrinker",
    estimate = round(exp(fitIV_g$est["everdrinker"]), 2),
    std.error = round(summary.g$coefficients[2], 2),
    statistic = round(summary.g$coefficients[3], 2),
    p.value =  round(summary.g$coefficients[4], 2),
    conf.low = round(exp(confint(fitIV_g)[1]), 2),
    conf.high = round(exp(confint(fitIV_g)[2]), 2),
    covariates = paste(covariates, collapse = " + ")
  )
  
  # Concatenate model results
  res <- rbind(res.ts, res.g)
  rownames(res) <- NULL
  
  # Return concatenated results
  return(res)
}

# IVGLM model for both ts and g-est with a binary exposure, continuous IV, and binary outcome, without covariates implemented with ivtools package
ivglm_model_no_covars <- function(data, ancestry, variant, outcome) {

  #two-stage estimation
  fitX.LZ <-
    glm(formula = as.formula(
      paste0(
        "everdrinker ~ ",
        add.backtick(variant)
      )
    ),
    family = "binomial",
    data = data)
  fitY.LX <-
    glm(formula = as.formula(
      paste0(
        outcome,
        " ~ + everdrinker"
      )
    ),
    family = "binomial",
    data = data)
  fitIV_ts  <-
    ivglm(
      estmethod = "ts",
      X = "everdrinker",
      Y = outcome,
      fitX.LZ = fitX.LZ,
      fitY.LX = fitY.LX,
      data = data,
      ctrl = T
    )
  
  # summarize ts model
  summary.ts <- summary(fitIV_ts)
    
  # export to results table
  res.ts <- data.frame(
    variant = variant,
    model = "ts",
    ancestry = ancestry,
    term = "everdrinker",
    estimate = round(exp(fitIV_ts$est["everdrinker"]), 2),
    std.error = round(summary.ts$coefficients[2,2], 2),
    statistic = round(summary.ts$coefficients[2,3], 2),
    p.value =  round(summary.ts$coefficients[2,4], 2),
    conf.low = round(exp(confint(fitIV_ts)[2,1]), 2),
    conf.high = round(exp(confint(fitIV_ts)[2,2]), 2),
    covariates = "none"
  )
  
  # G-estimation
  fitZ.L <-
    glm(formula = as.formula(
      paste0(
        add.backtick(variant),
        " ~ 1"
      )
    ),
    data = data)
  
  fitY.LZX <-
    glm(formula = as.formula(
      paste0(
        outcome,
        " ~ ",
        add.backtick(variant),
        " + everdrinker"
      )
    ),
    family = "binomial",
    data = data)
  
  fitIV_g  <-
    ivglm(
      estmethod = "g",
      X = "everdrinker",
      Y = outcome,
      fitZ.L = fitZ.L,
      fitY.LZX = fitY.LZX,
      data = data,
      link = "logit"
    )

  # Summarize g-est model
  summary.g <- summary(fitIV_g)
  
  # Export g-est model results
  res.g <- data.frame(
    variant = variant,
    model = "g",
    ancestry = ancestry,
    term = "everdrinker",
    estimate = round(exp(fitIV_g$est["everdrinker"]), 2),
    std.error = round(summary.g$coefficients[2], 2),
    statistic = round(summary.g$coefficients[3], 2),
    p.value =  round(summary.g$coefficients[4], 2),
    conf.low = round(exp(confint(fitIV_g)[1]), 2),
    conf.high = round(exp(confint(fitIV_g)[2]), 2),
    covariates = "none"
  )
  
  # Concatenate model results
  res <- rbind(res.ts, res.g)
  rownames(res) <- NULL
  
  # Return concatenated results
  return(res)
}
```


```{r}
# IVGLM model for both ts and g-est with continuous exposure, continuous IV, and binary outcome, implemented with ivtools package
ivglm_model_exp_num <- function(data, ancestry, variant, exposure, outcome, covariates) {

  ## Two-stage estimation
  fitX.LZ <-
    glm(formula = as.formula(
      paste0(
        exposure, " ~ ",
        add.backtick(variant),
        " + ", paste(covariates, collapse = " + ")
      )
    ),
    data = data)
  
  fitY.LX <-
    glm(formula = as.formula(
      paste0(
        outcome,
        " ~ ", exposure, " + ", paste(covariates, collapse = " + ")
      )
    ),
    family = "binomial",
    data = data)
  
  fitIV_ts  <-
    ivglm(
      estmethod = "ts",
      X = "everdrinker",
      Y = outcome,
      fitX.LZ = fitX.LZ,
      fitY.LX = fitY.LX,
      data = data,
      ctrl = T
    )
  
  # summarize ts model
  summary.ts <- summary(fitIV_ts)
    
  # export to results table
  res.ts <- data.frame(
    variant = variant,
    model = "ts",
    ancestry = ancestry,
    term = exposure,
    estimate = round(exp(fitIV_ts$est[exposure]), 2),
    std.error = round(summary.ts$coefficients[2,2], 2),
    statistic = round(summary.ts$coefficients[2,3], 2),
    p.value =  round(summary.ts$coefficients[2,4], 2),
    conf.low = round(exp(confint(fitIV_ts)[2,1]), 2),
    conf.high = round(exp(confint(fitIV_ts)[2,2]), 2),
    covariates = paste(covariates, collapse = " + ")
  )
  
  ## G-estimation
  fitZ.L <-
    glm(formula = as.formula(
      paste0(
        add.backtick(variant),
        " ~ ", paste(covariates, collapse = " + ")
      )
    ),
    data = data)
  
  fitY.LZX <-
    glm(formula = as.formula(
      paste0(
        outcome,
        " ~ ",
        add.backtick(variant),
        " + ", exposure, " + ", paste(covariates, collapse = " + ")
      )
    ),
    family = "binomial",
    data = data)
  
  fitIV_g  <-
    ivglm(
      estmethod = "g",
      X = exposure,
      Y = outcome,
      fitZ.L = fitZ.L,
      fitY.LZX = fitY.LZX,
      data = data,
      link = "logit"
    )

  # Summarize g-est model
  summary.g <- summary(fitIV_g)
  
  # Export g-est model results
  res.g <- data.frame(
    variant = variant,
    model = "g",
    ancestry = ancestry,
    term = exposure,
    estimate = round(exp(fitIV_g$est[exposure]), 2),
    std.error = round(summary.g$coefficients[2], 2),
    statistic = round(summary.g$coefficients[3], 2),
    p.value =  round(summary.g$coefficients[4], 2),
    conf.low = round(exp(confint(fitIV_g)[1]), 2),
    conf.high = round(exp(confint(fitIV_g)[2]), 2),
    covariates = paste(covariates, collapse = " + ")
  )
  
  # Concatenate model results
  res <- rbind(res.ts, res.g)
  rownames(res) <- NULL
  
  # Return concatenated results
  return(res)
}

# IVGLM model for both ts and g-est with continuous exposure, continuous IV, and binary outcome, implemented with ivtools package
ivglm_model_exp_num_no_covars <- function(data, ancestry, variant, exposure, outcome) {

  ## Two-stage estimation
  fitX.LZ <-
    glm(formula = as.formula(
      paste0(
        exposure, " ~ ",
        add.backtick(variant)
      )
    ),
    data = data)
  
  fitY.LX <-
    glm(formula = as.formula(
      paste0(
        outcome,
        " ~ ", exposure
      )
    ),
    family = "binomial",
    data = data)
  
  fitIV_ts  <-
    ivglm(
      estmethod = "ts",
      X = "everdrinker",
      Y = outcome,
      fitX.LZ = fitX.LZ,
      fitY.LX = fitY.LX,
      data = data,
      ctrl = T
    )
  
  # summarize ts model
  summary.ts <- summary(fitIV_ts)
    
  # export to results table
  res.ts <- data.frame(
    variant = variant,
    model = "ts",
    ancestry = ancestry,
    term = exposure,
    estimate = round(exp(fitIV_ts$est[exposure]), 2),
    std.error = round(summary.ts$coefficients[2,2], 2),
    statistic = round(summary.ts$coefficients[2,3], 2),
    p.value =  round(summary.ts$coefficients[2,4], 2),
    conf.low = round(exp(confint(fitIV_ts)[2,1]), 2),
    conf.high = round(exp(confint(fitIV_ts)[2,2]), 2),
    covariates = "none"
  )
  
  ## G-estimation
  fitZ.L <-
    glm(formula = as.formula(
      paste0(
        add.backtick(variant),
        " ~ ", paste(covariates, collapse = " + ")
      )
    ),
    data = data)
  
  fitY.LZX <-
    glm(formula = as.formula(
      paste0(
        outcome,
        " ~ ",
        add.backtick(variant),
        " + ", exposure, " + ", paste(covariates, collapse = " + ")
      )
    ),
    family = "binomial",
    data = data)
  
  fitIV_g  <-
    ivglm(
      estmethod = "g",
      X = exposure,
      Y = outcome,
      fitZ.L = fitZ.L,
      fitY.LZX = fitY.LZX,
      data = data,
      link = "logit"
    )

  # Summarize g-est model
  summary.g <- summary(fitIV_g)
  
  # Export g-est model results
  res.g <- data.frame(
    variant = variant,
    model = "g",
    ancestry = ancestry,
    term = exposure,
    estimate = round(exp(fitIV_g$est[exposure]), 2),
    std.error = round(summary.g$coefficients[2], 2),
    statistic = round(summary.g$coefficients[3], 2),
    p.value =  round(summary.g$coefficients[4], 2),
    conf.low = round(exp(confint(fitIV_g)[1]), 2),
    conf.high = round(exp(confint(fitIV_g)[2]), 2),
    covariates = "none"
  )
  
  # Concatenate model results
  res <- rbind(res.ts, res.g)
  rownames(res) <- NULL
  
  # Return concatenated results
  return(res)
}
```

# Analytic

# PGS of GSCAN DPW SNPs
Analytic 99 SNP 1SMR - EA
```{r}
these.covars <- c(ancestry.covars, precision.covariates)

ivglm_model(data = dat.ea, ancestry = "ea", variant = "pgs.norm", outcome = "AD12.Dementia.outcome", covariates = ancestry.covars)
ivglm_model(data = dat.ea, ancestry = "ea", variant = "pgs.norm", outcome = "AD12.CIND.outcome", covariates = ancestry.covars)
ivglm_model(data = dat.ea, ancestry = "ea", variant = "pgs.norm", outcome = "AD12.binary", covariates = ancestry.covars)

ivglm_model(data = dat.ea, ancestry = "ea", variant = "pgs.norm", outcome = "AD12.Dementia.outcome", covariates = covariates)
ivglm_model(data = dat.ea, ancestry = "ea", variant = "pgs.norm", outcome = "AD12.CIND.outcome", covariates = covariates)
ivglm_model(data = dat.ea, ancestry = "ea", variant = "pgs.norm", outcome = "AD12.binary", covariates = covariates)

ivglm_model(data = dat.ea, ancestry = "ea", variant = "pgs.norm", outcome = "AD12.Dementia.outcome", covariates = these.covars)
ivglm_model(data = dat.ea, ancestry = "ea", variant = "pgs.norm", outcome = "AD12.CIND.outcome", covariates = these.covars)
ivglm_model(data = dat.ea, ancestry = "ea", variant = "pgs.norm", outcome = "AD12.binary", covariates = these.covars)
```

# PGS of all SNPs

AAIC Poster/Abstract
```{r}
dat.test <- dat.ea %>% filter(NIWWAVE == T)
dat.test$AD12 %>% summary

ivglm_model(data = dat.ea %>% filter(NIWWAVE == T), ancestry = "ea", variant = "PGS4_DPW_GSCAN19", outcome = "AD12.Dementia.outcome", covariates = ancestry.covars)
ivglm_model(data = dat.ea %>% filter(NIWWAVE == T), ancestry = "ea", variant = "PGS4_DPW_GSCAN19", outcome = "AD12.CIND.outcome", covariates = ancestry.covars)
```


Analytic whole-genome SNP 1SMR - EA
```{r}
ivglm_model(data = dat.ea, ancestry = "ea", variant = "PGS4_DPW_GSCAN19", outcome = "AD12.Dementia.outcome", covariates = ancestry.covars)
ivglm_model(data = dat.ea, ancestry = "ea", variant = "PGS4_DPW_GSCAN19", outcome = "AD12.CIND.outcome", covariates = ancestry.covars)
ivglm_model(data = dat.ea, ancestry = "ea", variant = "PGS4_DPW_GSCAN19", outcome = "AD12.binary", covariates = ancestry.covars)

ivglm_model(data = dat.ea, ancestry = "ea", variant = "PGS4_DPW_GSCAN19", outcome = "AD12.Dementia.outcome", covariates = covariates)
ivglm_model(data = dat.ea, ancestry = "ea", variant = "PGS4_DPW_GSCAN19", outcome = "AD12.CIND.outcome", covariates = covariates)
ivglm_model(data = dat.ea, ancestry = "ea", variant = "PGS4_DPW_GSCAN19", outcome = "AD12.binary", covariates = covariates)

ivglm_model(data = dat.ea, ancestry = "ea", variant = "PGS4_DPW_GSCAN19", outcome = "AD12.Dementia.outcome", covariates = these.covars)
ivglm_model(data = dat.ea, ancestry = "ea", variant = "PGS4_DPW_GSCAN19", outcome = "AD12.CIND.outcome", covariates = these.covars)
ivglm_model(data = dat.ea, ancestry = "ea", variant = "PGS4_DPW_GSCAN19", outcome = "AD12.binary", covariates = these.covars)
```
Analytic 99 SNP 1SMR - AA
```{r}
ivglm_model(data = dat.aa, ancestry = "aa", variant = "pgs.norm", outcome = "AD12.Dementia.outcome", covariates = ancestry.covars)
ivglm_model(data = dat.aa, ancestry = "aa", variant = "pgs.norm", outcome = "AD12.CIND.outcome", covariates = ancestry.covars)
ivglm_model(data = dat.aa, ancestry = "aa", variant = "pgs.norm", outcome = "AD12.binary", covariates = ancestry.covars)

ivglm_model(data = dat.aa, ancestry = "aa", variant = "pgs.norm", outcome = "AD12.Dementia.outcome", covariates = these.covars)
ivglm_model(data = dat.aa, ancestry = "aa", variant = "pgs.norm", outcome = "AD12.CIND.outcome", covariates = these.covars)
ivglm_model(data = dat.aa, ancestry = "aa", variant = "pgs.norm", outcome = "AD12.binary", covariates = these.covars)
```
Analytic whole-genome SNP 1SMR - AA

```{r}
ivglm_model(data = dat.aa, ancestry = "aa", variant = "PGS4_DPW_GSCAN19", outcome = "AD12.Dementia.outcome", covariates = ancestry.covars)
ivglm_model(data = dat.aa, ancestry = "aa", variant = "PGS4_DPW_GSCAN19", outcome = "AD12.CIND.outcome", covariates = ancestry.covars)
ivglm_model(data = dat.aa, ancestry = "aa", variant = "PGS4_DPW_GSCAN19", outcome = "AD12.binary", covariates = ancestry.covars)

ivglm_model(data = dat.aa, ancestry = "aa", variant = "PGS4_DPW_GSCAN19", outcome = "AD12.Dementia.outcome", covariates = these.covars)
ivglm_model(data = dat.aa, ancestry = "aa", variant = "PGS4_DPW_GSCAN19", outcome = "AD12.CIND.outcome", covariates = these.covars)
ivglm_model(data = dat.aa, ancestry = "aa", variant = "PGS4_DPW_GSCAN19", outcome = "AD12.binary", covariates = these.covars)
```


```{r}
ivglm_model_no_covars(data = dat.ea, ancestry = "ea", variant = "PGS4_DPW_GSCAN19", outcome = "AD12.Dementia.outcome")
ivglm_model_no_covars(data = dat.ea, ancestry = "ea", variant = "PGS4_DPW_GSCAN19", outcome = "AD12.CIND.outcome")
ivglm_model_no_covars(data = dat.ea, ancestry = "ea", variant = "PGS4_DPW_GSCAN19", outcome = "AD12.binary")
```

# Drinks per week exposure

```{r}
ivglm_model_exp_num(data = dat.ea, ancestry = "ea", variant = "PGS4_DPW_GSCAN19", exposure = "drinkpweek", outcome = "AD12.Dementia.outcome", covariates = ancestry.covars)
ivglm_model_exp_num(data = dat.ea, ancestry = "ea", variant = "PGS4_DPW_GSCAN19", exposure = "drinkpweek", outcome = "AD12.CIND.outcome", covariates = ancestry.covars)
ivglm_model_exp_num(data = dat.ea, ancestry = "ea", variant = "PGS4_DPW_GSCAN19", exposure = "drinkpweek", outcome = "AD12.binary", covariates = ancestry.covars)
```

```{r}
ivglm_model_exp_num(data = dat.aa, ancestry = "aa", variant = "PGS4_DPW_GSCAN19", exposure = "drinkpweek", outcome = "AD12.Dementia.outcome", covariates = ancestry.covars)
ivglm_model_exp_num(data = dat.aa, ancestry = "aa", variant = "PGS4_DPW_GSCAN19", exposure = "drinkpweek", outcome = "AD12.CIND.outcome", covariates = ancestry.covars)
ivglm_model_exp_num(data = dat.aa, ancestry = "aa", variant = "PGS4_DPW_GSCAN19", exposure = "drinkpweek", outcome = "AD12.binary", covariates = ancestry.covars)
```

```{r}
hist(dat.ea$PGS4_DPW_GSCAN19)
hist(dat.ea$pgs.norm)
hist(dat.aa$pgs.norm)
```

```{r}
ivglm_model_exp_num_no_covars(data = dat.ea, ancestry = "ea", variant = "pgs.norm", exposure = "drinkpweek", outcome = "AD12.Dementia.outcome")
ivglm_model_exp_num_no_covars(data = dat.ea, ancestry = "ea", variant = "pgs.norm", exposure = "drinkpweek", outcome = "AD12.CIND.outcome")
ivglm_model_exp_num_no_covars(data = dat.ea, ancestry = "ea", variant = "pgs.norm", exposure = "drinkpweek", outcome = "AD12.binary")
```

```{r}
ivglm_model_exp_num(data = dat.ea, ancestry = "ea", variant = "pgs.norm", exposure = "drinkpweek", outcome = "AD12.Dementia.outcome", covariates = ancestry.covars)
ivglm_model_exp_num(data = dat.ea, ancestry = "ea", variant = "pgs.norm", exposure = "drinkpweek", outcome = "AD12.CIND.outcome", covariates = ancestry.covars)
ivglm_model_exp_num(data = dat.ea, ancestry = "ea", variant = "pgs.norm", exposure = "drinkpweek", outcome = "AD12.binary", covariates = ancestry.covars)
```

```{r}
ivglm_model_exp_num(data = dat.ea, ancestry = "ea", variant = "pgs.norm", exposure = "drinkpweek", outcome = "AD12.Dementia.outcome", covariates = covariates)
ivglm_model_exp_num(data = dat.ea, ancestry = "ea", variant = "pgs.norm", exposure = "drinkpweek", outcome = "AD12.CIND.outcome", covariates = covariates)
ivglm_model_exp_num(data = dat.ea, ancestry = "ea", variant = "pgs.norm", exposure = "drinkpweek", outcome = "AD12.binary", covariates = covariates)
```

Each variant individually with drinkpweek
```{r}
res <- map(colnames(variants), ~ ivglm_model_exp_num(data = dat.ea, ancestry = "ea", variant = .x, exposure = "drinkpweek", outcome = "AD12.Dementia.outcome", covariates = ancestry.covars)) %>% bind_rows()
```


# Each variant individually

```{r}
res <- map(colnames(variants), ~ ivglm_model(data = dat.ea, ancestry = "ea", variant = .x, outcome = "AD12.Dementia.outcome", covariates = covariates)) %>% bind_rows()
```

```{r}
res.base <- map(colnames(variants), ~ ivglm_model(data = dat.ea, ancestry = "ea", variant = .x, outcome = "AD12.Dementia.outcome", covariates = c("AncestryPC_1_5A", "AncestryPC_1_5B", "AncestryPC_1_5C", "AncestryPC_1_5D", "AncestryPC_1_5E"))) %>% bind_rows()
```

```{r}
res.ad12 <- map(colnames(variants), ~ ivglm_model(data = dat.ea, ancestry = "ea", variant = .x, outcome = "AD12.binary", covariates = covariates)) %>% bind_rows()
```

```{r}
res.ad12.base <- map(colnames(variants), ~ ivglm_model(data = dat.ea, ancestry = "ea", variant = .x, outcome = "AD12.binary", covariates = c("AncestryPC_1_5A", "AncestryPC_1_5B", "AncestryPC_1_5C", "AncestryPC_1_5D", "AncestryPC_1_5E"))) %>% bind_rows()
```


Univariate associations
```{r}
fit <- glm(formula = as.formula("AD12.Dementia.outcome ~ everdrinker"),
      family = "binomial", data = dat.ea)
summary(fit)

fit_variant_glm <- function(data, variant) {
  fit <- glm(formula = as.formula(paste0("AD12.Dementia.outcome ~ ", add.backtick(variant))),
      family = "binomial", data = data)
  return(tidy(fit))
}
res.fit_variant_glm <- map(colnames(variants), ~ fit_variant_glm(data = dat.ea, .x)) %>% bind_rows() %>% filter(!(term == "(Intercept)"))

fit_variant_glm_exposure <- function(data, variant) {
  fit <- glm(formula = as.formula(paste0("everdrinker ~ ", add.backtick(variant))),
      family = "binomial", data = data)
  return(tidy(fit))
}
res.fit_variant_glm_exposure <- map(colnames(variants), ~ fit_variant_glm_exposure(data = dat.ea, .x)) %>% bind_rows() %>% filter(!(term == "(Intercept)"))
```

```{r}
ex.form <-
  create.formula(outcome.name = "AD12.Dementia.outcome",
                 input.names = colnames(variants),
                 dat = dat.ea)

r2.fit <- glm(formula = ex.form$formula,
    family = "binomial",
    data = dat.ea)

summary(r2.fit)
```


### Bootstrap
From ivtools paper Appendix D
```{r}
set.seed(1)
```


```{r}
#variant <- "9:109345993:T:C"
#data <- dat.ea
#outcome <- "AD12.Dementia.outcome"

bootfun <- function(data, indicies) {
  
  variant = "9:109345993:T:C"
  outcome = "AD12.Dementia.outcome"
  
  # boot package input helper
  dd <- data[indicies,]
  
  #two-stage estimation
  fitX.LZ <-
    glm(formula = as.formula(
      paste0(
        "everdrinker ~ ",
        add.backtick(variant),
        " + GENDER + NAGE +  AncestryPC_1_5A + AncestryPC_1_5B + AncestryPC_1_5C + AncestryPC_1_5D + AncestryPC_1_5E"
      )
    ),
    family = "binomial",
    data = dd)
  fitY.LX <-
    glm(formula = as.formula(
      paste0(
        outcome,
        " ~ + everdrinker + GENDER + NAGE +  AncestryPC_1_5A + AncestryPC_1_5B + AncestryPC_1_5C + AncestryPC_1_5D + AncestryPC_1_5E"
      )
    ),
    family = "binomial",
    data = dd)
  fitIV_ts  <-
    ivglm(
      estmethod = "ts",
      X = "everdrinker",
      Y = outcome,
      fitX.LZ = fitX.LZ,
      fitY.LX = fitY.LX,
      data = dd,
      ctrl = T,
      vcov.fit = FALSE
    )
  est_ts <- fitIV_ts$est["everdrinker"]
  
  # Debugging purposes
  #est_ts
  
  # G-estimation
  fitZ.L <-
    glm(formula = as.formula(
      paste0(
        add.backtick(variant),
        " ~ GENDER + NAGE + AncestryPC_1_5A + AncestryPC_1_5B + AncestryPC_1_5C + AncestryPC_1_5D + AncestryPC_1_5E"
      )
    ),
    data = dd)
  
  fitY.LZX <-
    glm(formula = as.formula(
      paste0(
        outcome,
        " ~ ",
        add.backtick(variant),
        " + everdrinker + GENDER + NAGE +  AncestryPC_1_5A + AncestryPC_1_5B + AncestryPC_1_5C + AncestryPC_1_5D + AncestryPC_1_5E"
      )
    ),
    family = "binomial",
    data = dd)
  
  fitIV_g  <-
    ivglm(
      estmethod = "g",
      X = "everdrinker",
      Y = outcome,
      fitZ.L = fitZ.L,
      fitY.LZX = fitY.LZX,
      data = dd,
      link = "logit",
      vcov.fit = FALSE
    )
  est_g <- fitIV_g$est["everdrinker"]
  
  # Debugging purposes
  #est_g
  
  return(c(est_ts, est_g))
}
bb <- boot(data=dat.ea, statistic=bootfun, R=10000)
```


R = 100
```{r}
apply(bb$t, MARGIN=2, FUN=sd, na.rm=TRUE)
```

R = 1000
```{r}
apply(bb$t, MARGIN=2, FUN=sd, na.rm=TRUE)
```
R = 10000
```{r}
apply(bb$t, MARGIN=2, FUN=sd, na.rm=TRUE)
```





























Attempt to rerun previous analysis with everdrinker instead of drinksperweek as the outcome; Error in g %*% psi : requires numeric/complex matrix/vector arguments solved by converting input data to dataframe with as.data.frame()
```{r, eval=F}
fitZ.L <- glm(formula = EA_PGS4_DPW_GSCAN19 ~ 1,
              family = "gaussian", data = dat.ea)
summary(fitZ.L)

fitY.LZX <- glm(formula = AD12.Dementia.outcome ~ EA_PGS4_DPW_GSCAN19 + everdrinker,
                family = "binomial", data = dat.ea)
summary(fitY.LZX)

fitIV_g  <- ivglm(estmethod = "g", X = "everdrinker", Y = "AD12.Dementia.outcome", fitZ.L = fitZ.L, fitY.LZX = fitY.LZX,
                  data = dat.ea, link = "identity")

g.dem.ea.tidy <- tidy_ivglm_summary(fitIV_g)
g.dem.ea.tidy$outcome <- "Dementia"
g.dem.ea.tidy$sample <- "European ancestry"
g.dem.ea.tidy
```

```{r, eval=F}

fitZ.L   <- glm(formula = EA_PGS4_DPW_GSCAN19 ~ 1,
              family = "gaussian", data = dat.ea)
summary(fitZ.L)

fitY.LZX <- glm(formula = AD12.CIND.outcome ~ EA_PGS4_DPW_GSCAN19 + drinkpweek,
                family = "binomial", data = dat.ea)
summary(fitY.LZX)


fitIV_g  <- ivglm(estmethod = "g", X = "drinkpweek", fitZ.L = fitZ.L, fitY.LZX = fitY.LZX,
                  data = dat.ea, link = "logit")
summary(fitIV_g)

g.cind.ea.tidy <- tidy_ivglm_summary(fitIV_g)
g.cind.ea.tidy$outcome <- "CIND"
g.cind.ea.tidy$sample <- "European ancestry"
```